<h1 style="float:right"><a href="#" onclick='setPerspectiveName("Default")'><b>&larr;</b>Back</a></h1>
<img src="images/ChIPMunk_logo.png" align="right">
<center><h1>ChIPMunk for BioUML</h1></center>
<table>
<tr><td valign="top"><h3>Experienced user?</h3>Try ChIPMunk analyses directly:
<ul>
	<li><big><a href="#" onclick="doImportMultiple('ChIPMunk/User data/Data')">Import data</a></big><br>
	Load your BED files to supply to the ChIPMunk<br><br></li>
	<li><big><a href="#de=ChIPMunk/Analyses/ChIPMunk&docOnly=true">ChIPMunk</a></big><br>
	Create motifs from ChIP-Seq data<br><br></li>
	<li><big><a href="#de=ChIPMunk/Analyses/DiChIPMunk&docOnly=true">DiChIPMunk</a></big><br>
	Create di-nucleotide motifs from ChIP-Seq data<br><br></li>
	<li><big><a href="#de=ChIPMunk/Analyses/ChIPHorde&docOnly=true">ChIPHorde</a></big><br>
	Discover multiple significant motifs<br><br></li>
	<li><big><a href="#de=ChIPMunk/Analyses/DiChIPHorde&docOnly=true">DiChIPHorde</a></big><br>
	Discover multiple significant di-nucleotide motifs<br><br></li>
</ul>
<td valign="top"><h3>Novice user?</h3>Try simple workflows with a few parameters:
<ul>
	<li><big><a href="#de=ChIPMunk/Workflows/Reads to matrices&docOnly=true">Construct a matrix by the reads</a></big><br>
	Reads must be supplied as BED files (output of Bowtie, etc.)<br><br></li>
	<li><big><a href="#de=ChIPMunk/Workflows/Peaks to matrices&docOnly=true">Construct a matrix by the peaks</a></big><br>
	Peaks must be supplied as BED file (output of MACS, etc.)<br><br></li>
</ul>
</table>
<p>Visit <a target="_blank" href="http://autosome.ru/smbsm/librettos/libretto_chipmunk/chipmunk_home.rhtml">ChIPMunk web-site</a> for more info</p>
