<h2>BED format</h2>

<p>
  The BED format<span class="ref">1</span> provides a flexible way to define the data lines that
  are displayed in an annotation track. BED lines have three required fields and
  nine additional optional fields. The number of fields per line must be consistent
  throughout any single set of data in an annotation track. The order of the
  optional fields is binding: lower-numbered fields must always be populated if
  higher-numbered fields are used.
</p>

<p>
  The first three required BED fields are:
</p>
  
<ol>
  <li><b>chrom</b> - The name of the chromosome (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671).
  
  <li>
    <b>chromStart</b> - The starting position of the feature in the chromosome or scaffold.
    The first base in a chromosome is numbered 0.
  </li>
  
  <li>
    <b>chromEnd</b> - The ending position of the feature in the
    chromosome or scaffold. The <i>chromEnd</i> base is not included in the
    display of the feature. For example, the first 100 bases of a
    chromosome are defined as <i>chromStart=0, chromEnd=100</i>, and span
    the bases numbered 0-99.
  </li>

<p>The 9 additional optional BED fields are:</p>

  <li>
    <b>name</b> - Defines the name of the BED line.  This label is
    displayed to the left of the BED line in the Genome Browser
    window when the track is open to full display mode or directly to the
    left of the item in pack mode.
  </li>
  
  <li>
    <b>score</b> - A score between 0 and 1000.
  </li>
  
  <li><b>strand</b> - Defines the strand - as either '+' or '-'.
  
  <li>
    <b>thickStart</b> - The starting position at which the feature
    is drawn thickly (for example, the start codon in gene displays).
  </li>
  
  <li>
    <b>thickEnd</b> - The ending position at which the feature is
    drawn thickly (for example, the stop codon in gene displays).
  </li>
  
  <li>
    <b>itemRgb</b> - An RGB value of the form R,G,B (e.g. 255,0,0).
    If the track line <em>itemRgb</em> attribute is set to "On",
    this RBG value will determine the display color of the data contained
    in this BED line. NOTE: It is recommended that a simple color scheme
    (eight colors or less) be used with this attribute to avoid
    overwhelming the color resources of the Genome Browser and your
    Internet browser.
  </li>
  
  <li><b>blockCount</b> - The number of blocks (exons) in the BED line.
  
  <li>
    <b>blockSizes</b> - A comma-separated list of the block
    sizes. The number of items in this list should correspond to
    <i>blockCount</i>.
   
  <li>
    <b>blockStarts</b> - A comma-separated list of block starts.
    All of the <i>blockStart</i> positions should be calculated relative to
    <i>chromStart</i>. The number of items in this list should correspond to <i>blockCount</i>.
  </li>
</ol>

<h3>Example</h3>

<p>
  Here's an example of an annotation track that uses a complete BED definition:
</p>

<pre>
track name=pairedReads description="Clone Paired Reads" useScore=1
chr22 1000 5000 cloneA 960 + 1000 5000 0 2 567,488, 0,3512
chr22 2000 6000 cloneB 900 - 2000 6000 0 2 433,399, 0,3601
</pre>

<h3>References</h3>

<ol class="references">
  <li><a href="http://genome.ucsc.edu/FAQ/FAQformat.html">http://genome.ucsc.edu/FAQ/FAQformat.html</a>
</ol>