<h2>BreakDancer output format (*.ctx)</h2>

<p>
	These files produced by BreakDancer software.
	Each output file is tab-separated text file, which consists of the following columns:
</p>
<ol>
<li>Chromosome 1
<li>Position 1
<li>Orientation 1
<li>Chromosome 2
<li>Position 2
<li>Orientation 2
<li>Type of a SV
<li>Size of a SV
<li>Confidence Score
<li>Total number of supporting read pairs
<li>Total number of supporting read pairs from each map file
<li>Estimated allele frequency
<li>Software version
<li>The run parameters
</ol>
<p>
Columns 1-3 and 4-6 are used to specify the coordinates of the two SV breakpoints. The orientation is a string that records the number of reads mapped to the plus (+) or the minus (-) strand in the anchoring regions.
</p>
<p>Column 7 is the type of SV detected: DEL (deletions), INS (insertion), INV (inversion), ITX (intra-chromosomal translocation), CTX (inter-chromosomal translocation), and Unknown.
</p> 
<p>Column 8 is the size of the SV in bp.  It is meaningless for inter-chromosomal translocations.</p> 
<p>Column 9 is the confidence score associated with the prediction.</p> 
<p>Column 11 can be used to dissect the origin of the supporting read pairs, which is useful in pooled analysis.  For example, one may want to give SVs that are supported by more than one libraries higher confidence than those detected in only one library.  It can also be used to distinguish somatic events from the germline, i.e., those detected in only the tumor libraries versus those detected in both the tumor and the normal libraries.</p>
<p>Column 12 is currently a placeholder for displaying estimated allele frequency. The allele frequencies estimated in this version are not accurate and should not be trusted.</p>
<p>Column 13 and 14 are information useful to reproduce the results.</p>

<h3>Examples:</h3>
<pre>1 10000 10+0- 2 20000 7+10- CTX -296 99 10 tB|10 1.00 BreakDancerMax-0.0.1 t1</pre>
<p>An inter-chromosomal translocation that starts from chr1:10000 and goes into chr2:20000 with 10 supporting read pairs from the library tB and a confidence score of 99.</p>

<pre>1 59257 5+1- 1 60164 0+5- DEL 862 99 5 nA|2:tB|1 0.56 BreakDancerMax-0.0.1 c4</pre>
<p>A deletion between chr1:59257 and chr1:60164 connected by 5 read pairs, among which 2 in library nA and 1 in library tB support the deletion hypothesis. This deletion is detected by BreakDancerMax-0.0.1 with a separation threshold of 4 s.d.</p>

<pre>1 62767 10+0- 1 63126 0+10- INS -13 36 10 NA|10 1.00 BreakDancerMini-0.0.1 q10</pre>
<p>An 13 bp insertion detected by BreakDancerMini between chr1:62767 and chr1:63126 with 10 supporting read pairs from a single library 'NA' and a confidence score of 36.</p>
<h3>References</h3>
<ol>
<li><a href="http://sourceforge.net/projects/breakdancer/">BreakDancer SourceForge page</a>
<li><a href="https://github.com/kenchen/breakdancer">BreakDancer readme and format description</a>
</ol>