<html>
<body>
  <p>
    Success 128 of 303
  </p>
  <table width='100%' border='1'>
    <tr>
      <td>#</td>
      <td>Folder</td>
      <td>Method name</td>
      <td>Interpreter</td>
      <td>Status</td>
      <td>Time(ms)</td>
      <td>Count</td>
      <td>Compare</td>
      <td>Errors</td>
    </tr>
    <tr>
      <td>
        0
      </td>
      <td>
        EncodeTools
      </td>
      <td>
        gencode_partition1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/gencode_partition_out.bed
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        1
      </td>
      <td>
        EncodeTools
      </td>
      <td>
        random_intervals1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastq_quality_boxplot
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        3
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastq_quality_converter
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fastq_qual_conv1.out
File not exists: /tmp/fastq_qual_conv1a.out
File not exists: /tmp/fastq_qual_conv2.out
File not exists: /tmp/fastq_qual_conv2n.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        4
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastq_quality_filter
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fastq_qual_filter1a.out
File not exists: /tmp/fastq_qual_filter1b.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        5
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastq_to_fasta
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fastq_to_fasta1a.out
File not exists: /tmp/fastq_to_fasta1b.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        6
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_artifacts_filter
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fastx_artifacts1.out
File not exists: /tmp/fastx_artifacts2.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        7
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_barcode_splitter
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fastx_barcode_splitter1.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        8
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_clipper
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        9
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_collapser
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        10
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_nucleotides_distribution
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        11
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_quality_statistics
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fastq_stats1.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        12
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_renamer
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        13
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_reverse_complement
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fastx_reverse_complement1.out
File not exists: /tmp/fastx_reverse_complement2.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        14
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        cshl_fastx_trimmer
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fastx_trimmer1.out
File not exists: /tmp/fastx_trimmer2.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        15
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_combiner
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        7002
      </td>
      <td>
        6
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_combiner.py '$fasta_file' '${fasta_file.extension}' '$qual_file' '${qual_file.extension}' '$output_file' '$force_quality_encoding'
Traceback (most recent call last):
  File "fastq_combiner.py", line 49, in <module>
    main()
  File "fastq_combiner.py", line 25, in main
    out = fastqWriter( open( output_filename, 'wb' ), format = format, force_quality_encoding = force_quality_encoding )
IOError: [Errno 2] No such file or directory: "'/tmp/combine_phiX_out_1.fastqcssanger'"

File not exists: /tmp/combine_phiX_out_1.fastqcssanger
Command: 
fastq_combiner.py '$fasta_file' '${fasta_file.extension}' '$qual_file' '${qual_file.extension}' '$output_file' '$force_quality_encoding'
Traceback (most recent call last):
  File "fastq_combiner.py", line 49, in <module>
    main()
  File "fastq_combiner.py", line 25, in main
    out = fastqWriter( open( output_filename, 'wb' ), format = format, force_quality_encoding = force_quality_encoding )
IOError: [Errno 2] No such file or directory: "'/tmp/combine_phiX_out_2.fastqcssanger'"

File not exists: /tmp/combine_phiX_out_2.fastqcssanger
Command: 
fastq_combiner.py '$fasta_file' '${fasta_file.extension}' '$qual_file' '${qual_file.extension}' '$output_file' '$force_quality_encoding'
Traceback (most recent call last):
  File "fastq_combiner.py", line 49, in <module>
    main()
  File "fastq_combiner.py", line 25, in main
    out = fastqWriter( open( output_filename, 'wb' ), format = format, force_quality_encoding = force_quality_encoding )
IOError: [Errno 2] No such file or directory: "'/tmp/wrapping_as_sanger.fastqsanger'"

File not exists: /tmp/wrapping_as_sanger.fastqsanger
Command: 
fastq_combiner.py '$fasta_file' '${fasta_file.extension}' '$qual_file' '${qual_file.extension}' '$output_file' '$force_quality_encoding'
Traceback (most recent call last):
  File "fastq_combiner.py", line 49, in <module>
    main()
  File "fastq_combiner.py", line 25, in main
    out = fastqWriter( open( output_filename, 'wb' ), format = format, force_quality_encoding = force_quality_encoding )
IOError: [Errno 2] No such file or directory: "'/tmp/wrapping_as_sanger_decimal.fastqsanger'"

File not exists: /tmp/wrapping_as_sanger_decimal.fastqsanger
Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        16
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_filter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        6002
      </td>
      <td>
        8
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_filter.py $input_file $fastq_filter_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_filter.py", line 34, in <module>
    main()
  File "fastq_filter.py", line 21, in main
    for i, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 456, in __iter__
    yield self.next()
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 436, in next
    rval = fastqSequencingRead.get_class_by_format( self.format )()        
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 18, in get_class_by_format
    assert format in FASTQ_FORMATS, 'Unknown format type specified: %s' % format
AssertionError: Unknown format type specified: 'sanger'

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_filter.py $input_file $fastq_filter_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_filter.py", line 34, in <module>
    main()
  File "fastq_filter.py", line 21, in main
    for i, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 456, in __iter__
    yield self.next()
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 436, in next
    rval = fastqSequencingRead.get_class_by_format( self.format )()        
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 18, in get_class_by_format
    assert format in FASTQ_FORMATS, 'Unknown format type specified: %s' % format
AssertionError: Unknown format type specified: 'cssanger'

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_filter.py $input_file $fastq_filter_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_filter.py", line 34, in <module>
    main()
  File "fastq_filter.py", line 21, in main
    for i, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 456, in __iter__
    yield self.next()
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 436, in next
    rval = fastqSequencingRead.get_class_by_format( self.format )()        
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 18, in get_class_by_format
    assert format in FASTQ_FORMATS, 'Unknown format type specified: %s' % format
AssertionError: Unknown format type specified: 'sanger'

Command: 
fastq_filter.py $input_file $fastq_filter_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_filter.py", line 34, in <module>
    main()
  File "fastq_filter.py", line 21, in main
    for i, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 456, in __iter__
    yield self.next()
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 436, in next
    rval = fastqSequencingRead.get_class_by_format( self.format )()        
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 18, in get_class_by_format
    assert format in FASTQ_FORMATS, 'Unknown format type specified: %s' % format
AssertionError: Unknown format type specified: 'sanger'

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_filter.py $input_file $fastq_filter_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_filter.py", line 34, in <module>
    main()
  File "fastq_filter.py", line 21, in main
    for i, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 456, in __iter__
    yield self.next()
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 436, in next
    rval = fastqSequencingRead.get_class_by_format( self.format )()        
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 18, in get_class_by_format
    assert format in FASTQ_FORMATS, 'Unknown format type specified: %s' % format
AssertionError: Unknown format type specified: 'sanger'

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_filter.py $input_file $fastq_filter_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_filter.py", line 34, in <module>
    main()
  File "fastq_filter.py", line 21, in main
    for i, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 456, in __iter__
    yield self.next()
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 436, in next
    rval = fastqSequencingRead.get_class_by_format( self.format )()        
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 18, in get_class_by_format
    assert format in FASTQ_FORMATS, 'Unknown format type specified: %s' % format
AssertionError: Unknown format type specified: 'sanger'

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_filter.py $input_file $fastq_filter_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_filter.py", line 34, in <module>
    main()
  File "fastq_filter.py", line 21, in main
    for i, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 456, in __iter__
    yield self.next()
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 436, in next
    rval = fastqSequencingRead.get_class_by_format( self.format )()        
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 18, in get_class_by_format
    assert format in FASTQ_FORMATS, 'Unknown format type specified: %s' % format
AssertionError: Unknown format type specified: 'sanger'

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_filter.py $input_file $fastq_filter_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_filter.py", line 34, in <module>
    main()
  File "fastq_filter.py", line 21, in main
    for i, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 456, in __iter__
    yield self.next()
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 436, in next
    rval = fastqSequencingRead.get_class_by_format( self.format )()        
  File "/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy_utils/sequence/fastq.py", line 18, in get_class_by_format
    assert format in FASTQ_FORMATS, 'Unknown format type specified: %s' % format
AssertionError: Unknown format type specified: 'sanger'

</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        17
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_groomer
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        18004
      </td>
      <td>
        29
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        18
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_manipulation
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        8002
      </td>
      <td>
        13
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

File not exists: /tmp/fastq_trimmer_out1.fastqsanger
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

File not exists: /tmp/fastq_trimmer_out1.fastqsanger
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

File not exists: /tmp/sanger_full_range_rev_comp.fastqsanger
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

File not exists: /tmp/misc_dna_as_sanger_rev_comp_1.fastqsanger
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

File not exists: /tmp/misc_rna_as_sanger_rev_comp_1.fastqsanger
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

File not exists: /tmp/sanger_full_range_rev_comp_1_seq.fastqsanger
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

File not exists: /tmp/sanger_full_range_as_rna.fastqsanger
Command: 
fastq_manipulation.py $input_file $fastq_manipulation_file $output_file $output_file.files_path '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_manipulation.py", line 37, in <module>
    main()
  File "fastq_manipulation.py", line 18, in main
    fastq_manipulator = imp.load_module( 'fastq_manipulator', open( script_filename ), script_filename, ( '', 'r', imp.PY_SOURCE ) )
  File "/tmp/config_1", line 41
    left_column_offset = int( round( float( ${ manipulate_block['manipulation_type']['manipulation']['offset_type']['left_column_offset'] } ) / 100.0 * float( len( new_read ) ) ) )
                                            ^
SyntaxError: invalid syntax

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        19
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_masker_by_quality
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_masker_by_quality.py '$input_file' '$output_file' -f '${input_file.extension[len( 'fastq' ):]}' -s '${quality_score}' -c '${score_comparison}'
      #if $mask_type.value == 'lowercase'
      --lowercase
      #else
      -m '${mask_type}'
      #end if
  
Usage: fastq_masker_by_quality.py [options] input_file output_file

fastq_masker_by_quality.py: error: option -f: invalid choice: "'sanger'" (choose from 'sanger', 'solexa', 'illumina')

File not exists: /tmp/sanger_full_range_masked_N.fastqsanger
Command: 
fastq_masker_by_quality.py '$input_file' '$output_file' -f '${input_file.extension[len( 'fastq' ):]}' -s '${quality_score}' -c '${score_comparison}'
      #if $mask_type.value == 'lowercase'
      --lowercase
      #else
      -m '${mask_type}'
      #end if
  
Usage: fastq_masker_by_quality.py [options] input_file output_file

fastq_masker_by_quality.py: error: option -f: invalid choice: "'sanger'" (choose from 'sanger', 'solexa', 'illumina')

File not exists: /tmp/sanger_full_range_masked_lowercase.fastqsanger
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        20
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_paired_end_deinterlacer
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_paired_end_deinterlacer.py '$input_file' '${input_file.extension[len( 'fastq' ):]}' '$output1_pairs_file' '$output2_pairs_file' '$output1_singles_file' '$output2_singles_file'
Traceback (most recent call last):
  File "fastq_paired_end_deinterlacer.py", line 64, in <module>
    main()
  File "fastq_paired_end_deinterlacer.py", line 14, in main
    input       = fastqNamedReader( open( input_filename, 'rb' ), format = type  )
IOError: [Errno 2] No such file or directory: "'/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/paired_end_merged.fastqsanger'"

File not exists: /tmp/paired_end_2.fastqsanger
File not exists: /tmp/paired_end_2_singles.fastqsanger
File not exists: /tmp/paired_end_1_singles.fastqsanger
File not exists: /tmp/paired_end_1.fastqsanger
Command: 
fastq_paired_end_deinterlacer.py '$input_file' '${input_file.extension[len( 'fastq' ):]}' '$output1_pairs_file' '$output2_pairs_file' '$output1_singles_file' '$output2_singles_file'
Traceback (most recent call last):
  File "fastq_paired_end_deinterlacer.py", line 64, in <module>
    main()
  File "fastq_paired_end_deinterlacer.py", line 14, in main
    input       = fastqNamedReader( open( input_filename, 'rb' ), format = type  )
IOError: [Errno 2] No such file or directory: "'/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/paired_end_merged_errors.fastqsanger'"

File not exists: /tmp/paired_end_2_cleaned.fastqsanger
File not exists: /tmp/paired_end_2_cleaned_singles.fastqsanger
File not exists: /tmp/paired_end_1_cleaned_singles.fastqsanger
File not exists: /tmp/paired_end_1_cleaned.fastqsanger
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        21
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_paired_end_interlacer
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_paired_end_interlacer.py '$input1_file' '${input1_file.extension[len( 'fastq' ):]}' '$input2_file' '${input2_file.extension[len( 'fastq' ):]}' '$outfile_pairs' '$outfile_singles'
Traceback (most recent call last):
  File "fastq_paired_end_interlacer.py", line 59, in <module>
    main()
  File "fastq_paired_end_interlacer.py", line 19, in main
    out_pairs = fastqWriter( open( outfile_pairs, 'wb' ), format = type )
IOError: [Errno 2] No such file or directory: "'/tmp/paired_end_merged.fastqsanger'"

File not exists: /tmp/paired_end_merged_singles.fastqsanger
File not exists: /tmp/paired_end_merged.fastqsanger
Command: 
fastq_paired_end_interlacer.py '$input1_file' '${input1_file.extension[len( 'fastq' ):]}' '$input2_file' '${input2_file.extension[len( 'fastq' ):]}' '$outfile_pairs' '$outfile_singles'
Traceback (most recent call last):
  File "fastq_paired_end_interlacer.py", line 59, in <module>
    main()
  File "fastq_paired_end_interlacer.py", line 19, in main
    out_pairs = fastqWriter( open( outfile_pairs, 'wb' ), format = type )
IOError: [Errno 2] No such file or directory: "'/tmp/paired_end_merged_cleaned.fastqsanger'"

File not exists: /tmp/paired_end_merged_cleaned_singles.fastqsanger
File not exists: /tmp/paired_end_merged_cleaned.fastqsanger
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        22
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_paired_end_joiner
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_paired_end_joiner.py '$input1_file' '${input1_file.extension[len( 'fastq' ):]}' '$input2_file' '${input2_file.extension[len( 'fastq' ):]}' '$output_file'
Traceback (most recent call last):
  File "fastq_paired_end_joiner.py", line 38, in <module>
    main()
  File "fastq_paired_end_joiner.py", line 16, in main
    input2 = fastqNamedReader( open( input2_filename, 'rb' ), input2_type )
IOError: [Errno 2] No such file or directory: "'/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/split_pair_reads_2.fastqsanger'"

File not exists: /tmp/3.fastqsanger
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        23
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_paired_end_splitter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_paired_end_splitter.py '$input1_file' '${input1_file.extension[len( 'fastq' ):]}' '$output1_file' '$output2_file'
Traceback (most recent call last):
  File "fastq_paired_end_splitter.py", line 33, in <module>
    main()
  File "fastq_paired_end_splitter.py", line 13, in main
    out1 = fastqWriter( open( output1_filename, 'wb' ), format = input_type )
IOError: [Errno 2] No such file or directory: "'/tmp/split_pair_reads_1.fastqsanger'"

File not exists: /tmp/split_pair_reads_1.fastqsanger
File not exists: /tmp/split_pair_reads_2.fastqsanger
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        24
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_quality_trimmer
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        4001
      </td>
      <td>
        6
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_trimmer_by_quality.py '$input_file' '$output_file' -f '${input_file.extension[len( 'fastq' ):]}' -s '$window_size' 
    -t '$step_size' -e '$trim_ends' -a '$aggregation_action' -x '$exclude_count' -c '$score_comparison' -q '$quality_score' 
    #if $keep_zero_length.value:
        -k
    #end if
  
Usage: fastq_trimmer_by_quality.py [options] input_file output_file

fastq_trimmer_by_quality.py: error: option -f: invalid choice: "'sanger'" (choose from 'sanger', 'cssanger', 'solexa', 'illumina')

File not exists: /tmp/sanger_full_range_quality_trimmed_out_1.fastqsanger
Command: 
fastq_trimmer_by_quality.py '$input_file' '$output_file' -f '${input_file.extension[len( 'fastq' ):]}' -s '$window_size' 
    -t '$step_size' -e '$trim_ends' -a '$aggregation_action' -x '$exclude_count' -c '$score_comparison' -q '$quality_score' 
    #if $keep_zero_length.value:
        -k
    #end if
  
Usage: fastq_trimmer_by_quality.py [options] input_file output_file

fastq_trimmer_by_quality.py: error: option -f: invalid choice: "'sanger'" (choose from 'sanger', 'cssanger', 'solexa', 'illumina')

File not exists: /tmp/sanger_full_range_quality_trimmed_out_2.fastqsanger
Command: 
fastq_trimmer_by_quality.py '$input_file' '$output_file' -f '${input_file.extension[len( 'fastq' ):]}' -s '$window_size' 
    -t '$step_size' -e '$trim_ends' -a '$aggregation_action' -x '$exclude_count' -c '$score_comparison' -q '$quality_score' 
    #if $keep_zero_length.value:
        -k
    #end if
  
Usage: fastq_trimmer_by_quality.py [options] input_file output_file

fastq_trimmer_by_quality.py: error: option -f: invalid choice: "'sanger'" (choose from 'sanger', 'cssanger', 'solexa', 'illumina')

File not exists: /tmp/sanger_full_range_quality_trimmed_out_3.fastqsanger
Command: 
fastq_trimmer_by_quality.py '$input_file' '$output_file' -f '${input_file.extension[len( 'fastq' ):]}' -s '$window_size' 
    -t '$step_size' -e '$trim_ends' -a '$aggregation_action' -x '$exclude_count' -c '$score_comparison' -q '$quality_score' 
    #if $keep_zero_length.value:
        -k
    #end if
  
Usage: fastq_trimmer_by_quality.py [options] input_file output_file

fastq_trimmer_by_quality.py: error: option -f: invalid choice: "'sanger'" (choose from 'sanger', 'cssanger', 'solexa', 'illumina')

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
fastq_trimmer_by_quality.py '$input_file' '$output_file' -f '${input_file.extension[len( 'fastq' ):]}' -s '$window_size' 
    -t '$step_size' -e '$trim_ends' -a '$aggregation_action' -x '$exclude_count' -c '$score_comparison' -q '$quality_score' 
    #if $keep_zero_length.value:
        -k
    #end if
  
Usage: fastq_trimmer_by_quality.py [options] input_file output_file

fastq_trimmer_by_quality.py: error: option -f: invalid choice: "'sanger'" (choose from 'sanger', 'cssanger', 'solexa', 'illumina')

File not exists: /tmp/sanger_full_range_empty_reads.fastqsanger
Command: 
fastq_trimmer_by_quality.py '$input_file' '$output_file' -f '${input_file.extension[len( 'fastq' ):]}' -s '$window_size' 
    -t '$step_size' -e '$trim_ends' -a '$aggregation_action' -x '$exclude_count' -c '$score_comparison' -q '$quality_score' 
    #if $keep_zero_length.value:
        -k
    #end if
  
Usage: fastq_trimmer_by_quality.py [options] input_file output_file

fastq_trimmer_by_quality.py: error: option -f: invalid choice: "'sanger'" (choose from 'sanger', 'cssanger', 'solexa', 'illumina')

</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        25
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_stats
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_stats.py '$input_file' '$output_file' '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_stats.py", line 48, in <module>
    if __name__ == "__main__": main()
  File "fastq_stats.py", line 17, in main
    for num_reads, fastq_read in enumerate( fastqReader( open( input_filename ), format = input_type ) ):
IOError: [Errno 2] No such file or directory: "'/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/fastq_stats1.fastq'"

File not exists: /tmp/fastq_stats_1_out.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        26
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_to_fasta_python
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_to_fasta.py '$input_file' '$output_file' '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_fasta.py", line 22, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_fasta.py", line 13, in main
    out = fastaWriter( open( output_filename, 'wb' ) )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_fasta_python_1.out'"

File not exists: /tmp/fastq_to_fasta_python_1.out
Command: 
fastq_to_fasta.py '$input_file' '$output_file' '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_fasta.py", line 22, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_fasta.py", line 13, in main
    out = fastaWriter( open( output_filename, 'wb' ) )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_fasta_python_2.out'"

File not exists: /tmp/fastq_to_fasta_python_2.out
Command: 
fastq_to_fasta.py '$input_file' '$output_file' '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_fasta.py", line 22, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_fasta.py", line 13, in main
    out = fastaWriter( open( output_filename, 'wb' ) )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_fasta_python_1.out'"

File not exists: /tmp/fastq_to_fasta_python_1.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        27
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_to_tabular
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_to_tabular.py '$input_file' '$output_file' $descr_columns '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_tabular.py", line 38, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_tabular.py", line 21, in main
    out = open( output_filename, 'wb' )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_tabular_out_1.tabular'"

File not exists: /tmp/fastq_to_tabular_out_1.tabular
Command: 
fastq_to_tabular.py '$input_file' '$output_file' $descr_columns '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_tabular.py", line 38, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_tabular.py", line 21, in main
    out = open( output_filename, 'wb' )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_tabular_out_2.tabular'"

File not exists: /tmp/fastq_to_tabular_out_2.tabular
Command: 
fastq_to_tabular.py '$input_file' '$output_file' $descr_columns '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_tabular.py", line 38, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_tabular.py", line 21, in main
    out = open( output_filename, 'wb' )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_tabular_out_3.tabular'"

File not exists: /tmp/fastq_to_tabular_out_3.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        28
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastq_trimmer
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        5
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 92, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 92, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 92, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 92, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 92, in respond
TypeError: 'DictWrapper' object is not subscriptable

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        29
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        fastqc
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(100)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 99, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'ext'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        30
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        quality_score_distribution
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
short_reads_figure_score.py $input1 $output1 
Traceback (most recent call last):
  File "short_reads_figure_score.py", line 12, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/solexaScore.png
Command: 
short_reads_figure_score.py $input1 $output1 
Traceback (most recent call last):
  File "short_reads_figure_score.py", line 12, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/454Score.png
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        31
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        solid2fastq
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '@853_7_53/1' <> 'null'
Different lines: '@853_7_53/2' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        32
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        solid_qual_boxplot
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        33
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        solid_qual_stats
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
solid_qual_stats.py $input $output1
Traceback (most recent call last):
  File "solid_qual_stats.py", line 138, in <module>
    __main__()
  File "solid_qual_stats.py", line 24, in __main__
    fout = open(sys.argv[2].strip(),'r+w')
IOError: [Errno 2] No such file or directory: '/tmp/qualsolid.stats'

File not exists: /tmp/qualsolid.stats
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        34
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        tabular_to_fastq
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
tabular_to_fastq.py '$input_file' '$output_file' '$identifier' '$sequence' '$quality'
Traceback (most recent call last):
  File "tabular_to_fastq.py", line 29, in <module>
    if __name__ == "__main__": main()
  File "tabular_to_fastq.py", line 7, in main
    identifier_col = int( sys.argv[3] ) - 1
ValueError: invalid literal for int() with base 10: "'1'"

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
Command: 
tabular_to_fastq.py '$input_file' '$output_file' '$identifier' '$sequence' '$quality'
Traceback (most recent call last):
  File "tabular_to_fastq.py", line 29, in <module>
    if __name__ == "__main__": main()
  File "tabular_to_fastq.py", line 7, in main
    identifier_col = int( sys.argv[3] ) - 1
ValueError: invalid literal for int() with base 10: "'1'"

Different lines: '@FAKE0001 Original version has PHRED scores from 0 to 93 inclusive (in that order)' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        35
      </td>
      <td>
        NGS_QC
      </td>
      <td>
        trim_reads
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        36
      </td>
      <td>
        bxops
      </td>
      <td>
        Annotation_Profiler_0
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr22	30128507	31828507	uc003bnx.1_cds_2_0_chr22_29227_f	0	+	multiz17way	1700000	1' <> 'null'
Different lines: '#tableName	tableChromosomeCoverage	tableChromosomeCount	tableRegionCoverage	tableRegionCount	allIntervalCount	allIntervalSize	allCoverage	allTableRegionsOverlaped	allIntervalsOverlapingTable	nrIntervalCount	nrIntervalSize	nrCoverage	nrTableRegionsOverlaped	nrIntervalsOverlapingTable' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        37
      </td>
      <td>
        bxops
      </td>
      <td>
        flanking_features_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        14002
      </td>
      <td>
        7
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        38
      </td>
      <td>
        bxops
      </td>
      <td>
        get_flanks1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        39
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_basecoverage_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        40
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_cluster_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        6001
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        41
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_complement_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        5001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
gops_complement.py $input1 $output -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol} -l ${chromInfo} $allchroms
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_complement.py", line 34, in main
    in_fname, out_fname = args
ValueError: too many values to unpack


Complement regions.

usage: gops_complement.py in_file out_file
    -1, --cols1=N,N,N,N: Columns for chrom, start, end, strand in file
    -l, --lengths=N: Filename of .len file for species (chromosome lengths)
    -a, --all: Complement all chromosomes (Genome-wide complement)


File not exists: /tmp/gops_complement_out.bed
Command: 
gops_complement.py $input1 $output -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol} -l ${chromInfo} $allchroms
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_complement.py", line 34, in main
    in_fname, out_fname = args
ValueError: too many values to unpack


Complement regions.

usage: gops_complement.py in_file out_file
    -1, --cols1=N,N,N,N: Columns for chrom, start, end, strand in file
    -l, --lengths=N: Filename of .len file for species (chromosome lengths)
    -a, --all: Complement all chromosomes (Genome-wide complement)


File not exists: /tmp/gops_complement_out_diffCols.dat
Command: 
gops_complement.py $input1 $output -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol} -l ${chromInfo} $allchroms
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_complement.py", line 34, in main
    in_fname, out_fname = args
ValueError: too many values to unpack


Complement regions.

usage: gops_complement.py in_file out_file
    -1, --cols1=N,N,N,N: Columns for chrom, start, end, strand in file
    -l, --lengths=N: Filename of .len file for species (chromosome lengths)
    -a, --all: Complement all chromosomes (Genome-wide complement)


File not exists: /tmp/gops_complement_out2.bed
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        42
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_concat_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
gops_concat.py $input1 $input2 $output -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol} -2 ${input2.metadata.chromCol},${input2.metadata.startCol},${input2.metadata.endCol},${input2.metadata.strandCol} $sameformat
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_concat.py", line 38, in main
    in_file_1, in_file_2, out_fname = args
ValueError: too many values to unpack


Concatenate two bed files.  The concatenated files are returned in the
same format as the first.  If --sameformat is specified, then all
columns will be treated as the same, and all fields will be saved,
although the output will be trimmed to match the primary input.  In
addition, if --sameformat is specified, missing fields will be padded
with a period(.).

usage: gops_concat.py in_file_1 in_file_2 out_file
    -1, --cols1=N,N,N,N: Columns for chrom, start, end, strand in first file
    -2, --cols2=N,N,N,N: Columns for chrom, start, end, strand in second file
    -s, --sameformat: All files are precisely the same format.


File not exists: /tmp/gops_concat_out1.bed
Command: 
gops_concat.py $input1 $input2 $output -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol} -2 ${input2.metadata.chromCol},${input2.metadata.startCol},${input2.metadata.endCol},${input2.metadata.strandCol} $sameformat
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_concat.py", line 38, in main
    in_file_1, in_file_2, out_fname = args
ValueError: too many values to unpack


Concatenate two bed files.  The concatenated files are returned in the
same format as the first.  If --sameformat is specified, then all
columns will be treated as the same, and all fields will be saved,
although the output will be trimmed to match the primary input.  In
addition, if --sameformat is specified, missing fields will be padded
with a period(.).

usage: gops_concat.py in_file_1 in_file_2 out_file
    -1, --cols1=N,N,N,N: Columns for chrom, start, end, strand in first file
    -2, --cols2=N,N,N,N: Columns for chrom, start, end, strand in second file
    -s, --sameformat: All files are precisely the same format.


File not exists: /tmp/gops_concat_out2.bed
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        43
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_coverage_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        5001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr1	147962192	147962580	CCDS989.1_cds_0_0_chr1_147962193_r	0	-	388	1.0' <> 'chr1	147962192	147962580	CCDS989.1_cds_0_0_chr1_147962193_r	0	-	0	0.0'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        44
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_intersect_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        13002
      </td>
      <td>
        9
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr1	147962192	147962580	CCDS989.1_cds_0_0_chr1_147962193_r	0	-' <> 'null'
Command: 
gops_intersect.py 
      $input1 $input2 $output

      #if isinstance( $input1.datatype, $__app__.datatypes_registry.get_datatype_by_extension('gff').__class__):
        -1 1,4,5,7 --gff1
      #else:
        -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol}
      #end if

      #if isinstance( $input2.datatype, $__app__.datatypes_registry.get_datatype_by_extension('gff').__class__):
        -2 1,4,5,7 --gff2
      #else:
          -2 ${input2.metadata.chromCol},${input2.metadata.startCol},${input2.metadata.endCol},${input2.metadata.strandCol} 
      #end if

      -m $min $returntype
  
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_intersect.py", line 41, in main
    in_fname, in2_fname, out_fname = args
ValueError: too many values to unpack


Find regions of first interval file that overlap regions in a second interval file.
Interval files can either be BED or GFF format.

usage: gops_intersect.py interval_file_1 interval_file_2 out_file
    -1, --cols1=N,N,N,N: Columns for start, end, strand in first file
    -2, --cols2=N,N,N,N: Columns for start, end, strand in second file
    -m, --mincols=N: Require this much overlap (default 1bp)
    -p, --pieces: just print pieces of second set (after padding)
    -G, --gff1: input 1 is GFF format, meaning start and end coordinates are 1-based, closed interval
    -H, --gff2: input 2 is GFF format, meaning start and end coordinates are 1-based, closed interval


File not exists: /tmp/gops_intersect_p_diffCols.bed
Command: 
gops_intersect.py 
      $input1 $input2 $output

      #if isinstance( $input1.datatype, $__app__.datatypes_registry.get_datatype_by_extension('gff').__class__):
        -1 1,4,5,7 --gff1
      #else:
        -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol}
      #end if

      #if isinstance( $input2.datatype, $__app__.datatypes_registry.get_datatype_by_extension('gff').__class__):
        -2 1,4,5,7 --gff2
      #else:
          -2 ${input2.metadata.chromCol},${input2.metadata.startCol},${input2.metadata.endCol},${input2.metadata.strandCol} 
      #end if

      -m $min $returntype
  
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_intersect.py", line 41, in main
    in_fname, in2_fname, out_fname = args
ValueError: too many values to unpack


Find regions of first interval file that overlap regions in a second interval file.
Interval files can either be BED or GFF format.

usage: gops_intersect.py interval_file_1 interval_file_2 out_file
    -1, --cols1=N,N,N,N: Columns for start, end, strand in first file
    -2, --cols2=N,N,N,N: Columns for start, end, strand in second file
    -m, --mincols=N: Require this much overlap (default 1bp)
    -p, --pieces: just print pieces of second set (after padding)
    -G, --gff1: input 1 is GFF format, meaning start and end coordinates are 1-based, closed interval
    -H, --gff2: input 2 is GFF format, meaning start and end coordinates are 1-based, closed interval


File not exists: /tmp/gops_intersect_p_out.bed
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        45
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_join_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        9002
      </td>
      <td>
        6
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        46
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_merge_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        5001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
gops_merge.py $input1 $output -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol} $returntype
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_merge.py", line 32, in main
    in_fname, out_fname = args
ValueError: too many values to unpack


Merge overlaping regions.

usage: gops_merge.py in_file out_file
    -1, --cols1=N,N,N,N: Columns for start, end, strand in first file
    -m, --mincols=N: Require this much overlap (default 1bp)
    -3, --threecol: Output 3 column bed


File not exists: /tmp/gops-merge.dat
Command: 
gops_merge.py $input1 $output -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol} $returntype
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_merge.py", line 32, in main
    in_fname, out_fname = args
ValueError: too many values to unpack


Merge overlaping regions.

usage: gops_merge.py in_file out_file
    -1, --cols1=N,N,N,N: Columns for start, end, strand in first file
    -m, --mincols=N: Require this much overlap (default 1bp)
    -3, --threecol: Output 3 column bed


File not exists: /tmp/gops_merge_diffCols.dat
Command: 
gops_merge.py $input1 $output -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol} $returntype
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_merge.py", line 32, in main
    in_fname, out_fname = args
ValueError: too many values to unpack


Merge overlaping regions.

usage: gops_merge.py in_file out_file
    -1, --cols1=N,N,N,N: Columns for start, end, strand in first file
    -m, --mincols=N: Require this much overlap (default 1bp)
    -3, --threecol: Output 3 column bed


File not exists: /tmp/gops_merge_out2.bed
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        47
      </td>
      <td>
        bxops
      </td>
      <td>
        gops_subtract_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        9002
      </td>
      <td>
        6
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr11	1812377	1812407	CCDS7726.1_cds_0_0_chr11_1812378_f	0	+' <> 'chr1	147962192	147962580	CCDS989.1_cds_0_0_chr1_147962193_r	0	-'
Command: 
gops_subtract.py 
      $input1 $input2 $output

      #if isinstance( $input1.datatype, $__app__.datatypes_registry.get_datatype_by_extension('gff').__class__):
        -1 1,4,5,7 --gff1
      #else:
        -1 ${input1.metadata.chromCol},${input1.metadata.startCol},${input1.metadata.endCol},${input1.metadata.strandCol}
      #end if

      #if isinstance( $input2.datatype, $__app__.datatypes_registry.get_datatype_by_extension('gff').__class__):
        -2 1,4,5,7 --gff2
      #else:
          -2 ${input2.metadata.chromCol},${input2.metadata.startCol},${input2.metadata.endCol},${input2.metadata.strandCol} 
      #end if

      -m $min $returntype
  
Exception while parsing command line:
Traceback (most recent call last):
  File "gops_subtract.py", line 41, in main
    in_fname, in2_fname, out_fname = args
ValueError: too many values to unpack


Find regions of first interval file that do not overlap regions in a second
interval file. Interval files can either be BED or GFF format.

usage: gops_subtract.py interval_file_1 interval_file_2 out_file
    -1, --cols1=N,N,N,N: Columns for start, end, strand in first file
    -2, --cols2=N,N,N,N: Columns for start, end, strand in second file
    -m, --mincols=N: Require this much overlap (default 1bp)
    -p, --pieces: just print pieces of second set (after padding)
    -G, --gff1: input 1 is GFF format, meaning start and end coordinates are 1-based, closed interval
    -H, --gff2: input 2 is GFF format, meaning start and end coordinates are 1-based, closed interval


File not exists: /tmp/gops-subtract-p.dat
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        48
      </td>
      <td>
        clustal
      </td>
      <td>
        clustalw
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(5)
      </td>
      <td>
        <pre>Command: 
 
    rgClustalw.py -i "$input" -o "$output" -s "$out_order" -l "$outlog" -t "$outname" -d "$dnarna"
    #if   ($range.mode=="part")
-b "$range.seq_range_start" -e "$range.seq_range_end"
    #end if
    #if ($outcontrol.outform=="clustal")
-f "CLUSTAL"
    #if ($outcontrol.out_seqnos=="ON")
-q "ON"
    #end if
    #end if
    #if ($outcontrol.outform=="phylip")
-f "PHYLIP"
    #end if
    #if ($outcontrol.outform=="fasta")
-f "FASTA"
    #end if
   
Traceback (most recent call last):
  File "rgClustalw.py", line 55, in <module>
    assert os.path.isfile(opts.input)
AssertionError

File not exists: /tmp/rgClustal_testout.log
File not exists: /tmp/rgClustal_testout.fasta
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        49
      </td>
      <td>
        convert
      </td>
      <td>
        MAF_To_BED1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        50
      </td>
      <td>
        convert
      </td>
      <td>
        MAF_To_Fasta1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        51
      </td>
      <td>
        convert
      </td>
      <td>
        MAF_To_Interval1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 92, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'id'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        52
      </td>
      <td>
        convert
      </td>
      <td>
        Sff_extractor
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        29006
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre>Time out</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        53
      </td>
      <td>
        convert
      </td>
      <td>
        axt_to_concat_fasta
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        29007
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre>Time out</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        54
      </td>
      <td>
        convert
      </td>
      <td>
        axt_to_fasta
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        29006
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre>Time out</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        55
      </td>
      <td>
        convert
      </td>
      <td>
        axt_to_lav_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        56
      </td>
      <td>
        convert
      </td>
      <td>
        bed2gff1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        57
      </td>
      <td>
        convert
      </td>
      <td>
        bed_to_bigBed
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/7.bigbed
File not exists: /tmp/7.bigbed
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        58
      </td>
      <td>
        convert
      </td>
      <td>
        fasta2tab
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        4001
      </td>
      <td>
        6
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        59
      </td>
      <td>
        convert
      </td>
      <td>
        fastq_to_fasta_python
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
fastq_to_fasta.py '$input_file' '$output_file' '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_fasta.py", line 22, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_fasta.py", line 13, in main
    out = fastaWriter( open( output_filename, 'wb' ) )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_fasta_python_1.out'"

File not exists: /tmp/fastq_to_fasta_python_1.out
Command: 
fastq_to_fasta.py '$input_file' '$output_file' '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_fasta.py", line 22, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_fasta.py", line 13, in main
    out = fastaWriter( open( output_filename, 'wb' ) )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_fasta_python_2.out'"

File not exists: /tmp/fastq_to_fasta_python_2.out
Command: 
fastq_to_fasta.py '$input_file' '$output_file' '${input_file.extension[len( 'fastq' ):]}'
Traceback (most recent call last):
  File "fastq_to_fasta.py", line 22, in <module>
    if __name__ == "__main__": main()
  File "fastq_to_fasta.py", line 13, in main
    out = fastaWriter( open( output_filename, 'wb' ) )
IOError: [Errno 2] No such file or directory: "'/tmp/fastq_to_fasta_python_1.out'"

File not exists: /tmp/fastq_to_fasta_python_1.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        60
      </td>
      <td>
        convert
      </td>
      <td>
        gff2bed1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        61
      </td>
      <td>
        convert
      </td>
      <td>
        gtf2bedgraph
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        62
      </td>
      <td>
        convert
      </td>
      <td>
        lav_to_bed1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 89, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 246, in valueFromSearchList
    _raiseNotFoundException(key, searchList)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'bed_file1'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        63
      </td>
      <td>
        convert
      </td>
      <td>
        tab2fasta
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        64
      </td>
      <td>
        convert
      </td>
      <td>
        wig_to_bigWig
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/2.bigwig
File not exists: /tmp/2.bigwig
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        65
      </td>
      <td>
        convert
      </td>
      <td>
        wiggle2simple1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        4001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        66
      </td>
      <td>
        fasta_manipulation
      </td>
      <td>
        cshl_fasta_formatter
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/fasta_formatter1.out
File not exists: /tmp/fasta_formatter2.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        67
      </td>
      <td>
        fasta_manipulation
      </td>
      <td>
        cshl_fasta_nucleotides_changer
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        68
      </td>
      <td>
        fasta_manipulation
      </td>
      <td>
        cshl_fastx_collapser
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        69
      </td>
      <td>
        fasta_manipulation
      </td>
      <td>
        fasta2tab
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        4001
      </td>
      <td>
        6
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        70
      </td>
      <td>
        fasta_manipulation
      </td>
      <td>
        fasta_compute_length
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        71
      </td>
      <td>
        fasta_manipulation
      </td>
      <td>
        fasta_concatenate0
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        72
      </td>
      <td>
        fasta_manipulation
      </td>
      <td>
        fasta_filter_by_length
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        73
      </td>
      <td>
        fasta_manipulation
      </td>
      <td>
        tab2fasta
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        74
      </td>
      <td>
        features
      </td>
      <td>
        gene2exon1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr1	147962006	147962580	NM_005997	0	-' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        75
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        GeneBed_Maf_Fasta2
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/gene_bed_maf_to_fasta_out.fasta
Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 98, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'metadata'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        76
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        Interval2Maf1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 125, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'dbkey'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 89, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'dbkey'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 125, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'dbkey'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        77
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        Interval2Maf_pairwise1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 86, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'dbkey'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        78
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        Interval_Maf_Merged_Fasta2
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        4001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/interval_maf_to_merged_fasta_out3.fasta
File not exists: /tmp/interval_maf_to_merged_fasta_out.dat
Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 98, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'metadata'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        79
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        MAF_Limit_To_Species1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        80
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        MAF_Reverse_Complement_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        81
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        MAF_Thread_For_Species1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        82
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        MAF_filter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        83
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        MAF_split_blocks_by_species1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        84
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        maf_by_block_number1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        85
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        maf_limit_size1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        86
      </td>
      <td>
        fetchAlign
      </td>
      <td>
        maf_stats1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/maf_stats_interval_out.dat
File not exists: /tmp/maf_stats_summary_out.dat
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        87
      </td>
      <td>
        fetchSeq
      </td>
      <td>
        Extract genomic DNA 1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        10002
      </td>
      <td>
        7
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/extract_genomic_dna_out1.fasta
File not exists: /tmp/extract_genomic_dna_out2.fasta
File not exists: /tmp/extract_genomic_dna_out3.interval
File not exists: /tmp/extract_genomic_dna_out4.gff
File not exists: /tmp/extract_genomic_dna_out5.fasta
File not exists: /tmp/extract_genomic_dna_out6.fasta
File not exists: /tmp/extract_genomic_dna_out7.fasta
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        88
      </td>
      <td>
        filter
      </td>
      <td>
        Extract_features1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        89
      </td>
      <td>
        filter
      </td>
      <td>
        Filter1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        4001
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr22	30120223	30120265	CCDS13897.1_cds_0_0_chr22_30120224_f	0	+' <> 'null'
File not exists: /tmp/filter1_test2.bed
File not exists: /tmp/filter1_test3.sam
Different lines: 'chr22	30120223	30120265	CCDS13897.1_cds_0_0_chr22_30120224_f	0	+' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        90
      </td>
      <td>
        filter
      </td>
      <td>
        Grep1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr1	147962192	147962580	CCDS989.1_cds_0_0_chr1_147962193_r	0	-' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        91
      </td>
      <td>
        filter
      </td>
      <td>
        gff_filter_by_attribute
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'attribute_types'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'attribute_types'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        92
      </td>
      <td>
        filter
      </td>
      <td>
        gff_filter_by_feature_count
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/gff_filter_by_feature_count_out1.gff
File not exists: /tmp/gff_filter_by_feature_count_out2.gff
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        93
      </td>
      <td>
        filter
      </td>
      <td>
        gtf_filter_by_attribute_values_list
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        94
      </td>
      <td>
        filter
      </td>
      <td>
        sort1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 114, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'other_style'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 114, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'other_style'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        95
      </td>
      <td>
        gatk
      </td>
      <td>
        gatk_analyze_covariates
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        contains(0)
      </td>
      <td>
        <pre>Command: 
gatk_wrapper.py
   --stdout "${output_log}"
   --html_report_from_directory "${output_html}" "${output_html.files_path}"
   -p 'java 
    -jar "${GALAXY_DATA_INDEX_DIR}/shared/jars/gatk/AnalyzeCovariates.jar"
    -recalFile "${input_recal}"
    -outputDir "${output_html.files_path}"
    ##-log "${output_log}"
    ##-Rscript,--path_to_Rscript path_to_Rscript; on path is good enough
    -resources "${GALAXY_DATA_INDEX_DIR}/gatk/R"         
    #if $analysis_param_type.analysis_param_type_selector == "advanced":
        --ignoreQ "${analysis_param_type.ignore_q}"
        --numRG "${analysis_param_type.num_read_groups}"
        --max_quality_score "${analysis_param_type.max_quality_score}"
        --max_histogram_value "${analysis_param_type.max_histogram_value}"
         ${analysis_param_type.do_indel_quality}
    #end if
   '
  
Usage: gatk_wrapper.py [options]

gatk_wrapper.py: error: no such option: -j

File not exists: /tmp/gatk/gatk_analyze_covariates/gatk_analyze_covariates_out_1.log.contains
File not exists: /tmp/gatk/gatk_analyze_covariates/gatk_analyze_covariates_out_1.html
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        96
      </td>
      <td>
        gatk
      </td>
      <td>
        gatk_count_covariates
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        contains(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        97
      </td>
      <td>
        gatk
      </td>
      <td>
        gatk_indel_realigner
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        contains(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        98
      </td>
      <td>
        gatk
      </td>
      <td>
        gatk_realigner_target_creator
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        contains(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        99
      </td>
      <td>
        gatk
      </td>
      <td>
        gatk_table_recalibration
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        contains(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 94, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'ext'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        100
      </td>
      <td>
        gatk
      </td>
      <td>
        gatk_unified_genotyper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        contains(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        101
      </td>
      <td>
        gd
      </td>
      <td>
        gd_extract_flanking_dna
      </td>
      <td>
        python2.5&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/gd.extract_flanking_dna.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        102
      </td>
      <td>
        gd
      </td>
      <td>
        gd_extract_primers
      </td>
      <td>
        python2.5&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/gd.extract_primers.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        103
      </td>
      <td>
        gd
      </td>
      <td>
        gd_select_restriction_enzymes
      </td>
      <td>
        python2.5&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/gd.select_restriction_enzymes.wsf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        104
      </td>
      <td>
        gd
      </td>
      <td>
        gd_select_snps
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 102, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'ref'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        105
      </td>
      <td>
        getext
      </td>
      <td>
        biomart
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        106
      </td>
      <td>
        getext
      </td>
      <td>
        biomart_test
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        107
      </td>
      <td>
        getext
      </td>
      <td>
        bx_browser
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        108
      </td>
      <td>
        getext
      </td>
      <td>
        cbi_rice_mart
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        109
      </td>
      <td>
        getext
      </td>
      <td>
        encode_db1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        110
      </td>
      <td>
        getext
      </td>
      <td>
        epigraph_import
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        111
      </td>
      <td>
        getext
      </td>
      <td>
        epigraph_import_test
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        112
      </td>
      <td>
        getext
      </td>
      <td>
        eupathdb
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        113
      </td>
      <td>
        getext
      </td>
      <td>
        fix_errors
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        114
      </td>
      <td>
        getext
      </td>
      <td>
        flymine
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        115
      </td>
      <td>
        getext
      </td>
      <td>
        flymine_test
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        116
      </td>
      <td>
        getext
      </td>
      <td>
        gramenemart
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        117
      </td>
      <td>
        getext
      </td>
      <td>
        hbvar
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        118
      </td>
      <td>
        getext
      </td>
      <td>
        metabolicmine
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        119
      </td>
      <td>
        getext
      </td>
      <td>
        microbial_import1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        120
      </td>
      <td>
        getext
      </td>
      <td>
        modENCODEfly
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        121
      </td>
      <td>
        getext
      </td>
      <td>
        modENCODEworm
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        122
      </td>
      <td>
        getext
      </td>
      <td>
        modmine
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        123
      </td>
      <td>
        getext
      </td>
      <td>
        ratmine
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        124
      </td>
      <td>
        getext
      </td>
      <td>
        ucsc_table_direct1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        125
      </td>
      <td>
        getext
      </td>
      <td>
        ucsc_table_direct_archaea1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        126
      </td>
      <td>
        getext
      </td>
      <td>
        ucsc_table_direct_test1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        127
      </td>
      <td>
        getext
      </td>
      <td>
        upload1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        128
      </td>
      <td>
        getext
      </td>
      <td>
        wormbase
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        129
      </td>
      <td>
        getext
      </td>
      <td>
        wormbase_test
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        130
      </td>
      <td>
        getext
      </td>
      <td>
        yeastmine
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        131
      </td>
      <td>
        group
      </td>
      <td>
        Grouping1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 110, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'opcol'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        132
      </td>
      <td>
        group
      </td>
      <td>
        column_join
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        133
      </td>
      <td>
        group
      </td>
      <td>
        comp1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        134
      </td>
      <td>
        group
      </td>
      <td>
        join1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr11	1812377	1812407	CCDS7726.1_cds_0_0_chr11_1812378_f	0	+' <> 'chr12	38440094	38440321	CCDS8736.1_cds_0_0_chr12_38440095_r	0	-	chr12	38440094	38440321	NM_052885_cds_0_0_chr12_38440095_r	0	-'
Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        135
      </td>
      <td>
        group
      </td>
      <td>
        subtract_query1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
subtract_query.py $input1 $input2 $begin_col $end_col $output
Traceback (most recent call last):
  File "subtract_query.py", line 106, in <module>
    main()
  File "subtract_query.py", line 62, in main
    begin_col = int(begin_col)
ValueError: invalid literal for int() with base 10: 'c1'

File not exists: /tmp/subtract-query-3.dat
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        136
      </td>
      <td>
        hgv
      </td>
      <td>
        freebayes_wrapper
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(4)
      </td>
      <td>
        <pre>File not exists: /tmp/freebayes_out.vcf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        137
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_add_scores
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/add_scores_output1.interval
File not exists: /tmp/add_scores_output2.interval
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        138
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_beam
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        139
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_codingSnps
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    codingSnps.pl $input1 $input2 Galaxy build=${input1.metadata.dbkey} loc=${GALAXY_DATA_INDEX_DIR}/codingSnps.loc chr=${input1.metadata.chromCol} start=${input1.metadata.startCol} end=${input1.metadata.endCol} snp=$col1 > $out_file1
  
Failed to find sequence directory in locfile /home/tolstyh/projects/other/galaxy/galaxy-dist/tool-data/codingSnps.loc
sh: twoBitToFa: not found
Couldn't finish twoBitToFa on chr1 1541789 1541792, 

Different lines: 'chr1	1541789	1541790	C/T	ENST00000355826	Phe:Leu/Phe	15	T' <> 'null'
Command: 

    codingSnps.pl $input1 $input2 Galaxy build=${input1.metadata.dbkey} loc=${GALAXY_DATA_INDEX_DIR}/codingSnps.loc chr=${input1.metadata.chromCol} start=${input1.metadata.startCol} end=${input1.metadata.endCol} snp=$col1 > $out_file1
  
Failed to find sequence directory in locfile /home/tolstyh/projects/other/galaxy/galaxy-dist/tool-data/codingSnps.loc
sh: twoBitToFa: not found
Couldn't finish twoBitToFa on chr22 15452408 15452411, 

Different lines: 'chr22	15452410	15452411	G/T	uc002zlp.1	Pro:Pro/Thr	344	G' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        140
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_ctdBatch
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        141
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_david
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        142
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_funDo
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/funDo_output1.interval
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        143
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_gpass
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        144
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_hilbertvis
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        sim_size(7168)
      </td>
      <td>
        <pre>File not exists: /tmp/hvis_mkar_chr22.pdf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        145
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_ldtools
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/ldOutput1.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        146
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_linkToGProfile
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        147
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_lps
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/lps_arrhythmia_log.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        148
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_pass
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        149
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_sift
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/sift_variants_result.tab
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        150
      </td>
      <td>
        hgv
      </td>
      <td>
        hgv_snpFreq
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    snpFreq2.pl $input $group1_1 $group1_2 $group1_3 $group2_1 $group2_2 $group2_3 0.05 $output
  
Can't exec "R": No such file or directory at snpFreq2.pl line 100.
Couldn't call fisher.text, No such file or directory

File not exists: /tmp/snpFreqTestOut.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        151
      </td>
      <td>
        hybrid
      </td>
      <td>
        assembly_stats
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        152
      </td>
      <td>
        hybrid
      </td>
      <td>
        quake
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        153
      </td>
      <td>
        hybrid
      </td>
      <td>
        quake_pe
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        154
      </td>
      <td>
        hybrid
      </td>
      <td>
        smrtpipe_hybrid
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        155
      </td>
      <td>
        hybrid
      </td>
      <td>
        soap_denovo
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        156
      </td>
      <td>
        hyphy
      </td>
      <td>
        hgv_add_scores
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/add_scores_output1.interval
File not exists: /tmp/add_scores_output2.interval
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        157
      </td>
      <td>
        hyphy
      </td>
      <td>
        hgv_codingSnps
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    codingSnps.pl $input1 $input2 Galaxy build=${input1.metadata.dbkey} loc=${GALAXY_DATA_INDEX_DIR}/codingSnps.loc chr=${input1.metadata.chromCol} start=${input1.metadata.startCol} end=${input1.metadata.endCol} snp=$col1 > $out_file1
  
Failed to find sequence directory in locfile /home/tolstyh/projects/other/galaxy/galaxy-dist/tool-data/codingSnps.loc
sh: twoBitToFa: not found
Couldn't finish twoBitToFa on chr1 1541789 1541792, 

Different lines: 'chr1	1541789	1541790	C/T	ENST00000355826	Phe:Leu/Phe	15	T' <> 'null'
Command: 

    codingSnps.pl $input1 $input2 Galaxy build=${input1.metadata.dbkey} loc=${GALAXY_DATA_INDEX_DIR}/codingSnps.loc chr=${input1.metadata.chromCol} start=${input1.metadata.startCol} end=${input1.metadata.endCol} snp=$col1 > $out_file1
  
Failed to find sequence directory in locfile /home/tolstyh/projects/other/galaxy/galaxy-dist/tool-data/codingSnps.loc
sh: twoBitToFa: not found
Couldn't finish twoBitToFa on chr22 15452408 15452411, 

Different lines: 'chr22	15452410	15452411	G/T	uc002zlp.1	Pro:Pro/Thr	344	G' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        158
      </td>
      <td>
        hyphy
      </td>
      <td>
        mutate_snp_codon_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        159
      </td>
      <td>
        indel_analysis
      </td>
      <td>
        indel_analysis
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        160
      </td>
      <td>
        indel_analysis
      </td>
      <td>
        indel_sam2interval
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        161
      </td>
      <td>
        indel_analysis
      </td>
      <td>
        indel_table
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 114, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'sum'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        162
      </td>
      <td>
        indel_analysis
      </td>
      <td>
        sam_indel_filter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        163
      </td>
      <td>
        liftOver
      </td>
      <td>
        liftOver1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/5_liftover_mapped.bed
File not exists: /tmp/5_liftover_unmapped.bed
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        164
      </td>
      <td>
        motifs
      </td>
      <td>
        meme_meme
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(8)
      </td>
      <td>
        <pre>File not exists: /tmp/meme/meme/meme_output_txt_1.txt
File not exists: /tmp/meme/meme/meme_output_xml_1.xml
File not exists: /tmp/meme/meme/meme_output_html_1.html
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        165
      </td>
      <td>
        motifs
      </td>
      <td>
        rgweblogo3
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        sim_size(10000)
      </td>
      <td>
        <pre>File not exists: /tmp/rgWebLogo3_test.jpg
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        166
      </td>
      <td>
        multReg
      </td>
      <td>
        BestSubsetsRegression1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        167
      </td>
      <td>
        multReg
      </td>
      <td>
        LinearRegression1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    linear_regression.py 
      $input1
      $response_col
      $predictor_cols
      $out_file1
      $out_file2
      1>/dev/null
  
Traceback (most recent call last):
  File "linear_regression.py", line 5, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/regr_out.tabular
File not exists: /tmp/regr_out.pdf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        168
      </td>
      <td>
        multReg
      </td>
      <td>
        rcve1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    rcve.py 
      $input1
      $response_col
      $predictor_cols
      $out_file1
      1>/dev/null
  
Traceback (most recent call last):
  File "rcve.py", line 6, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/rcve_out.dat
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        169
      </td>
      <td>
        multVar
      </td>
      <td>
        cca1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    cca.py 
      $input1
      $x_cols
      $y_cols
      $x_scale
      $y_scale
      $std_scores
      $out_file1
      $out_file2
  
Traceback (most recent call last):
  File "cca.py", line 5, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/cca_out1.tabular
File not exists: /tmp/cca_out2.pdf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        170
      </td>
      <td>
        multVar
      </td>
      <td>
        kcca1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        re_match(0)
      </td>
      <td>
        <pre>Command: 

    kcca.py 
      --input=$input1
      --output1=$out_file1
      --x_cols=$x_cols
      --y_cols=$y_cols
      --kernel=$kernelChoice.kernel
      --features=$features
      #if $kernelChoice.kernel == "rbfdot" or $kernelChoice.kernel == "anovadot":
      --sigma=$kernelChoice.sigma
      --degree="None"
      --scale="None"
      --offset="None"
      --order="None"
      #elif $kernelChoice.kernel == "polydot":
      --sigma="None"
      --degree=$kernelChoice.degree
      --scale=$kernelChoice.scale
      --offset=$kernelChoice.offset
      --order="None"
      #elif $kernelChoice.kernel == "tanhdot":
      --sigma="None"
      --degree="None"
      --scale=$kernelChoice.scale
      --offset=$kernelChoice.offset
      --order="None"
      #elif $kernelChoice.kernel == "besseldot":
      --sigma=$kernelChoice.sigma
      --degree=$kernelChoice.degree
      --scale="None"
      --offset="None"
      --order=$kernelChoice.order
      #elif $kernelChoice.kernel == "anovadot":
      --sigma=$kernelChoice.sigma
      --degree=$kernelChoice.degree
      --scale="None"
      --offset="None"
      --order="None"
      #else:
      --sigma="None"
      --degree="None"
      --scale="None"
      --offset="None"
      --order="None"
      #end if
  
Traceback (most recent call last):
  File "kcca.py", line 24, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/kcca_out1.tabular
Command: 

    kcca.py 
      --input=$input1
      --output1=$out_file1
      --x_cols=$x_cols
      --y_cols=$y_cols
      --kernel=$kernelChoice.kernel
      --features=$features
      #if $kernelChoice.kernel == "rbfdot" or $kernelChoice.kernel == "anovadot":
      --sigma=$kernelChoice.sigma
      --degree="None"
      --scale="None"
      --offset="None"
      --order="None"
      #elif $kernelChoice.kernel == "polydot":
      --sigma="None"
      --degree=$kernelChoice.degree
      --scale=$kernelChoice.scale
      --offset=$kernelChoice.offset
      --order="None"
      #elif $kernelChoice.kernel == "tanhdot":
      --sigma="None"
      --degree="None"
      --scale=$kernelChoice.scale
      --offset=$kernelChoice.offset
      --order="None"
      #elif $kernelChoice.kernel == "besseldot":
      --sigma=$kernelChoice.sigma
      --degree=$kernelChoice.degree
      --scale="None"
      --offset="None"
      --order=$kernelChoice.order
      #elif $kernelChoice.kernel == "anovadot":
      --sigma=$kernelChoice.sigma
      --degree=$kernelChoice.degree
      --scale="None"
      --offset="None"
      --order="None"
      #else:
      --sigma="None"
      --degree="None"
      --scale="None"
      --offset="None"
      --order="None"
      #end if
  
Traceback (most recent call last):
  File "kcca.py", line 24, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/kcca_out2.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        171
      </td>
      <td>
        multVar
      </td>
      <td>
        kpca1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    kpca.py 
      --input=$input1
      --output1=$out_file1
      --output2=$out_file2
      --var_cols=$var_cols
      --kernel=$kernelChoice.kernel
      --features=$features
      #if $kernelChoice.kernel == "rbfdot" or $kernelChoice.kernel == "anovadot":
      --sigma=$kernelChoice.sigma
      --degree="None"
      --scale="None"
      --offset="None"
      --order="None"
      #elif $kernelChoice.kernel == "polydot":
      --sigma="None"
      --degree=$kernelChoice.degree
      --scale=$kernelChoice.scale
      --offset=$kernelChoice.offset
      --order="None"
      #elif $kernelChoice.kernel == "tanhdot":
      --sigma="None"
      --degree="None"
      --scale=$kernelChoice.scale
      --offset=$kernelChoice.offset
      --order="None"
      #elif $kernelChoice.kernel == "besseldot":
      --sigma=$kernelChoice.sigma
      --degree=$kernelChoice.degree
      --scale="None"
      --offset="None"
      --order=$kernelChoice.order
      #elif $kernelChoice.kernel == "anovadot":
      --sigma=$kernelChoice.sigma
      --degree=$kernelChoice.degree
      --scale="None"
      --offset="None"
      --order="None"
      #else:
      --sigma="None"
      --degree="None"
      --scale="None"
      --offset="None"
      --order="None"
      #end if
  
Traceback (most recent call last):
  File "kpca.py", line 24, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/kpca_out1.tabular
File not exists: /tmp/kpca_out2.pdf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        172
      </td>
      <td>
        multVar
      </td>
      <td>
        pca1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    pca.py 
      $input1
      $var_cols
      $methodChoice.method
      $out_file1
      $out_file2
      #if $methodChoice.method == "svd":
      $methodChoice.scale
      #end if
  
Traceback (most recent call last):
  File "pca.py", line 5, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/pca_out1.tabular
File not exists: /tmp/pca_out2.pdf
Command: 

    pca.py 
      $input1
      $var_cols
      $methodChoice.method
      $out_file1
      $out_file2
      #if $methodChoice.method == "svd":
      $methodChoice.scale
      #end if
  
Traceback (most recent call last):
  File "pca.py", line 5, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/pca_out3.tabular
File not exists: /tmp/pca_out4.pdf
Command: 

    pca.py 
      $input1
      $var_cols
      $methodChoice.method
      $out_file1
      $out_file2
      #if $methodChoice.method == "svd":
      $methodChoice.scale
      #end if
  
Traceback (most recent call last):
  File "pca.py", line 5, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/pca_out5.tabular
File not exists: /tmp/pca_out6.pdf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        173
      </td>
      <td>
        ncbi_blast_plus_tools
      </td>
      <td>
        blastxml_to_tabular
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        6001
      </td>
      <td>
        8
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        174
      </td>
      <td>
        ncbi_blast_plus_tools
      </td>
      <td>
        ncbi_blastn_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        175
      </td>
      <td>
        ncbi_blast_plus_tools
      </td>
      <td>
        ncbi_blastp_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
blastp
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-task $blast_type
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
blastp -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_proteins.fasta" -task blastp -evalue 1e-8 -out /tmp/blastp_four_human_vs_rhodopsin.xml -outfmt 5 -num_threads 8 False -matrix BLOSUM62 True

[Errno 2] No such file or directory

File not exists: /tmp/blastp_four_human_vs_rhodopsin.xml
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
blastp
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-task $blast_type
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
blastp -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_proteins.fasta" -task blastp -evalue 1e-8 -out /tmp/blastp_four_human_vs_rhodopsin.tabular -outfmt 6 -num_threads 8 False -matrix BLOSUM62 True

[Errno 2] No such file or directory

File not exists: /tmp/blastp_four_human_vs_rhodopsin.tabular
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
blastp
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-task $blast_type
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
blastp -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_proteins.fasta" -task blastp -evalue 1e-8 -out /tmp/blastp_four_human_vs_rhodopsin_ext.tabular -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen" -num_threads 8 False -matrix BLOSUM62 True

[Errno 2] No such file or directory

File not exists: /tmp/blastp_four_human_vs_rhodopsin_ext.tabular
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
blastp
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-task $blast_type
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
blastp -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -task blastp -evalue 1e-8 -out /tmp/blastp_rhodopsin_vs_four_human.tabular -outfmt 6 -num_threads 8

[Errno 2] No such file or directory

File not exists: /tmp/blastp_rhodopsin_vs_four_human.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        176
      </td>
      <td>
        ncbi_blast_plus_tools
      </td>
      <td>
        ncbi_blastx_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
blastx
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
$adv_opts.strand
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
blastx -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_nucs.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -evalue 1e-10 -out /tmp/blastx_rhodopsin_vs_four_human.xml -outfmt 5 -num_threads 8

[Errno 2] No such file or directory

File not exists: /tmp/blastx_rhodopsin_vs_four_human.xml
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
blastx
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
$adv_opts.strand
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
blastx -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_nucs.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -evalue 1e-10 -out /tmp/blastx_rhodopsin_vs_four_human.tabular -outfmt 6 -num_threads 8

[Errno 2] No such file or directory

File not exists: /tmp/blastx_rhodopsin_vs_four_human.tabular
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
blastx
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
$adv_opts.strand
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
blastx -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_nucs.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -evalue 1e-10 -out /tmp/blastx_rhodopsin_vs_four_human_ext.tabular -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen" -num_threads 8

[Errno 2] No such file or directory

File not exists: /tmp/blastx_rhodopsin_vs_four_human_ext.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        177
      </td>
      <td>
        ncbi_blast_plus_tools
      </td>
      <td>
        ncbi_tblastn_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        4001
      </td>
      <td>
        5
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
tblastn
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
tblastn -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_nucs.fasta" -evalue 1e-10 -out /tmp/tblastn_four_human_vs_rhodopsin.xml -outfmt 5 -num_threads 8 false -matrix BLOSUM80 false

[Errno 2] No such file or directory

File not exists: /tmp/tblastn_four_human_vs_rhodopsin.xml
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
tblastn
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
tblastn -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_nucs.fasta" -evalue 1e-10 -out /tmp/tblastn_four_human_vs_rhodopsin_ext.tabular -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen" -num_threads 8 false -matrix BLOSUM80 false

[Errno 2] No such file or directory

File not exists: /tmp/tblastn_four_human_vs_rhodopsin_ext.tabular
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
tblastn
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
tblastn -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_nucs.fasta" -evalue 1e-10 -out /tmp/tblastn_four_human_vs_rhodopsin.tabular -outfmt 6 -num_threads 8 false -matrix BLOSUM80 false

[Errno 2] No such file or directory

File not exists: /tmp/tblastn_four_human_vs_rhodopsin.tabular
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
tblastn
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
tblastn -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_nucs.fasta" -evalue 1e-10 -out /tmp/tblastn_four_human_vs_rhodopsin.tabular -outfmt 6 -num_threads 8 false -matrix BLOSUM80 true

[Errno 2] No such file or directory

File not exists: /tmp/tblastn_four_human_vs_rhodopsin.tabular
Command: 
hide_stderr.py
## The command is a Cheetah template which allows some Python based syntax.
## Lines starting hash hash are comments. Galaxy will turn newlines into spaces
tblastn
-query "$query"
#if $db_opts.db_opts_selector == "db":
  -db "${db_opts.database.fields.path}"
#else:
  -subject "$db_opts.subject"
#end if
-evalue $evalue_cutoff
-out $output1
##Set the extended list here so if/when we add things, saved workflows are not affected
#if str($out_format)=="ext":
    -outfmt "6 std sallseqid score nident positive gaps ppos qframe sframe qseq sseq qlen slen"
#else:
    -outfmt $out_format
#end if
-num_threads 8
#if $adv_opts.adv_opts_selector=="advanced":
$adv_opts.filter_query
-matrix $adv_opts.matrix
## Need int(str(...)) because $adv_opts.max_hits is an InputValueWrapper object not a string
## Note -max_target_seqs overrides -num_descriptions and -num_alignments
#if (str($adv_opts.max_hits) and int(str($adv_opts.max_hits)) > 0):
-max_target_seqs $adv_opts.max_hits
#end if
#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
-word_size $adv_opts.word_size
#end if
##Ungapped disabled for now - see comments below
##$adv_opts.ungapped
$adv_opts.parse_deflines
## End of advanced options:
#end if
    
Error invoking command:
tblastn -query "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/four_human_proteins.fasta" -subject "/home/tolstyh/projects/other/galaxy/galaxy-dist/test-data/rhodopsin_nucs.fasta" -evalue 1e-10 -out /tmp/tblastn_four_human_vs_rhodopsin.html -outfmt 0 -html -num_threads 8 false -matrix BLOSUM80 false

[Errno 2] No such file or directory

File not exists: /tmp/tblastn_four_human_vs_rhodopsin.html
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        178
      </td>
      <td>
        ncbi_blast_plus_tools
      </td>
      <td>
        ncbi_tblastx_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        179
      </td>
      <td>
        ngs-rna-tools
      </td>
      <td>
        cuffcompare
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

        cuffcompare_wrapper.py 
            
            ## Use annotation reference?
            #if $annotation.use_ref_annotation == "Yes":
                -r $annotation.reference_annotation
                #if $annotation.ignore_nonoverlapping_reference:
                    -R
                #end if
            #end if
            
            ## Use sequence data?
            #if $seq_data.use_seq_data == "Yes":
	        -s
                #if $seq_data.seq_source.index_source == "history":
                    --ref_file=$seq_data.seq_source.ref_file
                #else:
                    --ref_file="None"
                #end if
                --dbkey=${first_input.metadata.dbkey} 
                --index_dir=${GALAXY_DATA_INDEX_DIR}
            #end if
            
            ## Outputs.
            --combined-transcripts=${transcripts_combined}
            
            ## Inputs.
            ${first_input}
            #for $input_file in $input_files:
              ${input_file.additional_input}
            #end for
            
    
Traceback (most recent call last):
  File "cuffcompare_wrapper.py", line 131, in <module>
    if __name__=="__main__": __main__()
  File "cuffcompare_wrapper.py", line 90, in __main__
    os.symlink( arg, input_file_name )
OSError: [Errno 17] File exists

File not exists: /tmp/cuffcompare_out3.tmap
File not exists: /tmp/cuffcompare_out5.gtf
File not exists: /tmp/cuffcompare_out7.txt
File not exists: /tmp/cuffcompare_out4.refmap
File not exists: /tmp/cuffcompare_out2.refmap
File not exists: /tmp/cuffcompare_out6.tracking
File not exists: /tmp/cuffcompare_out1.tmap
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        180
      </td>
      <td>
        ngs-rna-tools
      </td>
      <td>
        cuffdiff
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(200)
      </td>
      <td>
        <pre>Error: java.lang.NullPointerException
'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        181
      </td>
      <td>
        ngs-rna-tools
      </td>
      <td>
        cufflinks
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

        cufflinks_wrapper.py 
            --input=$input
            --assembled-isoforms-output=$assembled_isoforms
            --num-threads="4"
            -I $max_intron_len
            -F $min_isoform_fraction
            -j $pre_mrna_fraction
            
            ## Include reference annotation?
            #if $reference_annotation.use_ref == "Yes":
                -G $reference_annotation.reference_annotation_file
            #end if
            
            ## Set paired-end parameters?
            #if $singlePaired.sPaired == "Yes":
                -m $singlePaired.mean_inner_distance
                -s $singlePaired.inner_distance_std_dev
                #end if

            ## Normalization?
            #if str($do_normalization) == "Yes":
            -N
            #end if
            
            ## Bias correction?
            #if $bias_correction.do_bias_correction == "Yes":
	        -b
                #if $bias_correction.seq_source.index_source == "history":
                    --ref_file=$bias_correction.seq_source.ref_file
                #else:
                    --ref_file="None"
                #end if
                --dbkey=${input.metadata.dbkey} 
                --index_dir=${GALAXY_DATA_INDEX_DIR}
            #end if
    
Traceback (most recent call last):
  File "cufflinks_wrapper.py", line 152, in <module>
    if __name__=="__main__": __main__()
  File "cufflinks_wrapper.py", line 102, in __main__
    cmd += ( " -p %i" % int ( options.num_threads ) )
ValueError: invalid literal for int() with base 10: '"4"'

File not exists: /tmp/cufflinks_out2.fpkm_tracking
File not exists: /tmp/cufflinks_out3.fpkm_tracking
File not exists: /tmp/cufflinks_out1.gtf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        182
      </td>
      <td>
        ngs-rna-tools
      </td>
      <td>
        filter_combined_via_tracking
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        183
      </td>
      <td>
        ngs-rna-tools
      </td>
      <td>
        tophat
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        4
      </td>
      <td>
        sim_size(0)
      </td>
      <td>
        <pre>Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        184
      </td>
      <td>
        ngs-simulation
      </td>
      <td>
        ngs_simulation
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        185
      </td>
      <td>
        peak_calling
      </td>
      <td>
        bed2genetrack
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        186
      </td>
      <td>
        peak_calling
      </td>
      <td>
        peakcalling_ccat
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 96, in respond
TypeError: 'DictWrapper' object is not subscriptable

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        187
      </td>
      <td>
        peak_calling
      </td>
      <td>
        peakcalling_macs
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        re_match(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 92, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 246, in valueFromSearchList
    _raiseNotFoundException(key, searchList)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'output_extra_files'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 92, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 246, in valueFromSearchList
    _raiseNotFoundException(key, searchList)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'output_extra_files'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        188
      </td>
      <td>
        peak_calling
      </td>
      <td>
        predict2genetrack
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        189
      </td>
      <td>
        picard_beta
      </td>
      <td>
        PicardASMetrics
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(50)
      </td>
      <td>
        <pre>File not exists: /tmp/picard_output_alignment_summary_metrics.html
Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 125, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 238, in valueForName
    _wrapNotFoundException(e, fullName=name, namespace=obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 236, in valueForName
    return _valueForName(obj, name, executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'tool_data_tables' while searching for 'tool_data_tables'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        190
      </td>
      <td>
        picard_beta
      </td>
      <td>
        PicardGCBiasMetrics
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(50)
      </td>
      <td>
        <pre>Command: 

    picard_wrapper.py -i "$input_file" -d "$html_file.files_path" -t "$html_file"
    --windowsize "$windowsize" --mingenomefrac "$mingenomefrac" -n "$out_prefix" --tmpdir "${__new_file_path__}"
    -j ${GALAXY_DATA_INDEX_DIR}/shared/jars/CollectGcBiasMetrics.jar
#if $genomeSource.refGenomeSource == "history":
 --ref-file "$genomeSource.ownFile"
#else:
 --ref "${ filter( lambda x: str( x[0] ) == str( $genomeSource.index ), $__app__.tool_data_tables[ 'all_fasta' ].get_fields() )[0][-1] }"
#end if
  
Traceback (most recent call last):
  File "picard_wrapper.py", line 733, in <module>
    if __name__=="__main__": __main__()
  File "picard_wrapper.py", line 554, in __main__
    assert os.path.isfile(ref_file_name),'PicardGC needs a reference sequence - cannot read %s' % ref_file_name
AssertionError: PicardGC needs a reference sequence - cannot read /home/tolstyh/projects/other/galaxy/galaxy-dist/tools/picard/"/tmp/galaxy"/CollectGcBiasMetricss3J3dL.fasta

File not exists: /tmp/picard_output_GcBias_uploaded_hg18_summary_alignment_stats.html
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        191
      </td>
      <td>
        picard_beta
      </td>
      <td>
        PicardHsMetrics
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(212)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 97, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'ext'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        192
      </td>
      <td>
        picard_beta
      </td>
      <td>
        PicardInsertSize
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        contains(40)
      </td>
      <td>
        <pre>Command: 

   picard_wrapper.py -i "$input_file" -n "$out_prefix" --tmpdir "${__new_file_path__}" --taillimit "$tailLimit"
   --histwidth "$histWidth" --minpct "$minPct"
   -j "${GALAXY_DATA_INDEX_DIR}/shared/jars/CollectInsertSizeMetrics.jar" -d "$html_file.files_path" -t "$html_file"
  
Traceback (most recent call last):
  File "picard_wrapper.py", line 733, in <module>
    if __name__=="__main__": __main__()
  File "picard_wrapper.py", line 600, in __main__
    if float( opts.minpct) > 0.0:
ValueError: could not convert string to float: "0.01"

File not exists: /tmp/picard_output_insertsize_tinysam.html
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        193
      </td>
      <td>
        picard_beta
      </td>
      <td>
        picard_ARRG
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '@HD	VN:1.0	SO:coordinate' <> 'null'
Different lines: '@HD	VN:1.0	SO:coordinate' <> 'null'
Different lines: '      BC FJ@'FS7&i6b`F=xD7_SM' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        194
      </td>
      <td>
        picard_beta
      </td>
      <td>
        picard_BamIndexStats
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        contains(12)
      </td>
      <td>
        <pre>File not exists: /tmp/picard_BIS_output1.txt
File not exists: /tmp/picard_BIS_output2.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        195
      </td>
      <td>
        picard_beta
      </td>
      <td>
        picard_ReorderSam
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(12)
      </td>
      <td>
        <pre>Unsupported comparator: containsError: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 94, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 238, in valueForName
    _wrapNotFoundException(e, fullName=name, namespace=obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 236, in valueForName
    return _valueForName(obj, name, executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'tool_data_tables' while searching for 'tool_data_tables'

'Different lines: 'HWI-EAS91_1_30788AAXX:1:1:1598:534/1	0	phiX2	597	255	50M	*	0	0	GCGCTCTAATCTCTGGGCATCTGGCTATGATGTTGATGGAACTGACCAAA	IIIIIII""IIIIIIIIIIIIIIIIIIIIIIIIII"IIIIIII000IIII	XA:i:0	MD:Z:50	NM:i:0' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        196
      </td>
      <td>
        picard_beta
      </td>
      <td>
        picard_ReplaceSamHeader
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '@HD	VN:1.0	SO:coordinate' <> 'null'
Different lines: '@HD	VN:1.0	SO:coordinate' <> 'null'
Different lines: '      BC Mo03#mu^@jIMEI.!aIv	>$' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        197
      </td>
      <td>
        picard_beta
      </td>
      <td>
        rgEstLibComp
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(30)
      </td>
      <td>
        <pre>Command: 

   picard_wrapper.py -i "$input_file" -n "$out_prefix" --tmpdir "${__new_file_path__}" --minid "$minIDbases"
   --maxdiff "$maxDiff" --minmeanq "$minMeanQ" --readregex "$readRegex" --optdupdist "$optDupeDist"
   -j "${GALAXY_DATA_INDEX_DIR}/shared/jars/EstimateLibraryComplexity.jar" -d "$html_file.files_path" -t "$html_file"
  
Traceback (most recent call last):
  File "picard_wrapper.py", line 733, in <module>
    if __name__=="__main__": __main__()
  File "picard_wrapper.py", line 512, in __main__
    if float(opts.minid) > 0:
ValueError: could not convert string to float: "5"

File not exists: /tmp/picard_output_estlibcomplexity_tinysam.html
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        198
      </td>
      <td>
        picard_beta
      </td>
      <td>
        rgPicFixMate
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/picard_output_fixmate_sorted_pair.bam
File not exists: /tmp/picard_output_fixmate_sorted_pair.sam
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        199
      </td>
      <td>
        picard_beta
      </td>
      <td>
        rgPicardMarkDups
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(75)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 122, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'ext'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 122, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'ext'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        200
      </td>
      <td>
        plots
      </td>
      <td>
        XY_Plot_1
      </td>
      <td>
        bash&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/XY_Plot_1_out.pdf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        201
      </td>
      <td>
        plots
      </td>
      <td>
        barchart_gnuplot
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        202
      </td>
      <td>
        plots
      </td>
      <td>
        build_ucsc_custom_track_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 94, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'file_name'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        203
      </td>
      <td>
        plots
      </td>
      <td>
        gmaj_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        204
      </td>
      <td>
        plots
      </td>
      <td>
        histogram_rpy
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
histogram.py $input $out_file1 $numerical_column "$title" "$xlab" $breaks $density $frequency
Traceback (most recent call last):
  File "histogram.py", line 5, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/histogram_out1.pdf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        205
      </td>
      <td>
        plots
      </td>
      <td>
        laj_1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        206
      </td>
      <td>
        plots
      </td>
      <td>
        mutation_visualize
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 114, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'a_col'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 114, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'a_col'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        207
      </td>
      <td>
        plots
      </td>
      <td>
        qual_stats_boxplot
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/boxplot_summary_statistics_out.png
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        208
      </td>
      <td>
        plots
      </td>
      <td>
        scatterplot_rpy
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        209
      </td>
      <td>
        plots
      </td>
      <td>
        vcf_to_maf_customtrack1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        210
      </td>
      <td>
        regVar
      </td>
      <td>
        categorize_elements_satisfying_criteria
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        211
      </td>
      <td>
        regVar
      </td>
      <td>
        compute_motif_frequencies_for_all_motifs
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        6001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        212
      </td>
      <td>
        regVar
      </td>
      <td>
        compute_motifs_frequency
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        23004
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        213
      </td>
      <td>
        regVar
      </td>
      <td>
        cpgFilter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        214
      </td>
      <td>
        regVar
      </td>
      <td>
        delete_overlapping_indels
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        215
      </td>
      <td>
        regVar
      </td>
      <td>
        draw_stacked_barplots
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '%PDF-1.4' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        216
      </td>
      <td>
        regVar
      </td>
      <td>
        featureCoverage1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        217
      </td>
      <td>
        regVar
      </td>
      <td>
        getIndels_2way
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        218
      </td>
      <td>
        regVar
      </td>
      <td>
        indelRates_3way
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '#Block	hg18_InsRate	panTro2_InsRate	canFam2_InsRate	hg18_DelRate	panTro2_DelRate	canFam2_DelRate' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        219
      </td>
      <td>
        regVar
      </td>
      <td>
        indels_3way
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        220
      </td>
      <td>
        regVar
      </td>
      <td>
        microsatellite_birthdeath
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '#hg18chrom	hg18start	hg18end	hg18motif	hg18motifsize	hg18threshold	ponAbe2	ponAbe2.chr	ponAbe2.start	ponAbe2.end	ponAbe2.motif	ponAbe2.motifsize	ponAbe2.microsat	ponAbe2.mutation	ponAbe2.mutation.position	ponAbe2.mutation.from	ponAbe2.mutation.to	ponAbe2.insertion.details	ponAbe2.deletion.details	hg18	hg18.chr	hg18.start	hg18.end	hg18.motif	hg18.motifsize	hg18.microsat	hg18.mutation	hg18.mutation.position	hg18.mutation.from	hg18.mutation.to	hg18.insertion.details	hg18.deletion.details	panTro2	panTro2.chr	panTro2.start	panTro2.end	panTro2.motif	panTro2.motifsize	panTro2.microsat	panTro2.mutation	panTro2.mutation.position	panTro2.mutation.from	panTro2.mutation.to	panTro2.insertion.details	panTro2.deletion.details	rheMac2	rheMac2.chr	rheMac2.start	rheMac2.end	rheMac2.motif	rheMac2.motifsize	rheMac2.microsat	rheMac2.mutation	rheMac2.mutation.position	rheMac2.mutation.from	rheMac2.mutation.to	rheMac2.insertion.details	rheMac2.deletion.details	calJac1	calJac1.chr	calJac1.start	calJac1.end	calJac1.motif	calJac1.motifsize	calJac1.microsat	calJac1.mutation	calJac1.mutation.position	calJac1.mutation.from	calJac1.mutation.to	calJac1.insertion.details	calJac1.deletion.details	somechr	somestart	somened	event	source	hg18panTro2ponAbe2.insertions_num	hg18panTro2ponAbe2.deletions_num	hg18panTro2ponAbe2.motinsertions_num	hg18panTro2ponAbe2.motinsertionsf_num	hg18panTro2ponAbe2.motdeletions_num	hg18panTro2ponAbe2.motdeletionsf_num	hg18panTro2ponAbe2.noninsertions_num	hg18panTro2ponAbe2.nondeletions_num	hg18panTro2ponAbe2.binsertions_num	hg18panTro2ponAbe2.bdeletions_num	hg18panTro2ponAbe2.bmotinsertions_num	hg18panTro2ponAbe2.bmotinsertionsf_num	hg18panTro2ponAbe2.bmotdeletions_num	hg18panTro2ponAbe2.bmotdeletionsf_num	hg18panTro2ponAbe2.bnoninsertions_num	hg18panTro2ponAbe2.bnondeletions_num	hg18panTro2ponAbe2.dinsertions_num	hg18panTro2ponAbe2.ddeletions_num	hg18panTro2ponAbe2.dmotinsertions_num	hg18panTro2ponAbe2.dmotinsertionsf_num	hg18panTro2ponAbe2.dmotdeletions_num	hg18panTro2ponAbe2.dmotdeletionsf_num	hg18panTro2ponAbe2.dnoninsertions_num	hg18panTro2ponAbe2.dnondeletions_num	hg18panTro2ponAbe2.ninsertions_num	hg18panTro2ponAbe2.ndeletions_num	hg18panTro2ponAbe2.nmotinsertions_num	hg18panTro2ponAbe2.nmotinsertionsf_num	hg18panTro2ponAbe2.nmotdeletions_num	hg18panTro2ponAbe2.nmotdeletionsf_num	hg18panTro2ponAbe2.nnoninsertions_num	hg18panTro2ponAbe2.nnondeletions_num	hg18panTro2ponAbe2.substitutions_num	hg18panTro2ponAbe2.bsubstitutions_num	hg18panTro2ponAbe2.dsubstitutions_num	hg18panTro2ponAbe2.nsubstitutions_num	hg18panTro2ponAbe2.indels_num	hg18panTro2ponAbe2.subs_num	rhecalJac1ac2.insertions_num	rhecalJac1ac2.deletions_num	rhecalJac1ac2.motinsertions_num	rhecalJac1ac2.motinsertionsf_num	rhecalJac1ac2.motdeletions_num	rhecalJac1ac2.motdeletionsf_num	rhecalJac1ac2.noninsertions_num	rhecalJac1ac2.nondeletions_num	rhecalJac1ac2.binsertions_num	rhecalJac1ac2.bdeletions_num	rhecalJac1ac2.bmotinsertions_num	rhecalJac1ac2.bmotinsertionsf_num	rhecalJac1ac2.bmotdeletions_num	rhecalJac1ac2.bmotdeletionsf_num	rhecalJac1ac2.bnoninsertions_num	rhecalJac1ac2.bnondeletions_num	rhecalJac1ac2.dinsertions_num	rhecalJac1ac2.ddeletions_num	rhecalJac1ac2.dmotinsertions_num	rhecalJac1ac2.dmotinsertionsf_num	rhecalJac1ac2.dmotdeletions_num	rhecalJac1ac2.dmotdeletionsf_num	rhecalJac1ac2.dnoninsertions_num	rhecalJac1ac2.dnondeletions_num	rhecalJac1ac2.ninsertions_num	rhecalJac1ac2.ndeletions_num	rhecalJac1ac2.nmotinsertions_num	rhecalJac1ac2.nmotinsertionsf_num	rhecalJac1ac2.nmotdeletions_num	rhecalJac1ac2.nmotdeletionsf_num	rhecalJac1ac2.nnoninsertions_num	rhecalJac1ac2.nnondeletions_num	rhecalJac1ac2.substitutions_num	rhecalJac1ac2.bsubstitutions_num	rhecalJac1ac2.dsubstitutions_num	rhecalJac1ac2.nsubstitutions_num	rhecalJac1ac2.indels_num	rhecalJac1ac2.subs_num	hg18panTro2.insertions_num	hg18panTro2.deletions_num	hg18panTro2.motinsertions_num	hg18panTro2.motinsertionsf_num	hg18panTro2.motdeletions_num	hg18panTro2.motdeletionsf_num	hg18panTro2.noninsertions_num	hg18panTro2.nondeletions_num	hg18panTro2.binsertions_num	hg18panTro2.bdeletions_num	hg18panTro2.bmotinsertions_num	hg18panTro2.bmotinsertionsf_num	hg18panTro2.bmotdeletions_num	hg18panTro2.bmotdeletionsf_num	hg18panTro2.bnoninsertions_num	hg18panTro2.bnondeletions_num	hg18panTro2.dinsertions_num	hg18panTro2.ddeletions_num	hg18panTro2.dmotinsertions_num	hg18panTro2.dmotinsertionsf_num	hg18panTro2.dmotdeletions_num	hg18panTro2.dmotdeletionsf_num	hg18panTro2.dnoninsertions_num	hg18panTro2.dnondeletions_num	hg18panTro2.ninsertions_num	hg18panTro2.ndeletions_num	hg18panTro2.nmotinsertions_num	hg18panTro2.nmotinsertionsf_num	hg18panTro2.nmotdeletions_num	hg18panTro2.nmotdeletionsf_num	hg18panTro2.nnoninsertions_num	hg18panTro2.nnondeletions_num	hg18panTro2.substitutions_num	hg18panTro2.bsubstitutions_num	hg18panTro2.dsubstitutions_num	hg18panTro2.nsubstitutions_num	hg18panTro2.indels_num	hg18panTro2.subs_num	ponAbe2.insertions_num	ponAbe2.deletions_num	ponAbe2.motinsertions_num	ponAbe2.motinsertionsf_num	ponAbe2.motdeletions_num	ponAbe2.motdeletionsf_num	ponAbe2.noninsertions_num	ponAbe2.nondeletions_num	ponAbe2.binsertions_num	ponAbe2.bdeletions_num	ponAbe2.bmotinsertions_num	ponAbe2.bmotinsertionsf_num	ponAbe2.bmotdeletions_num	ponAbe2.bmotdeletionsf_num	ponAbe2.bnoninsertions_num	ponAbe2.bnondeletions_num	ponAbe2.dinsertions_num	ponAbe2.ddeletions_num	ponAbe2.dmotinsertions_num	ponAbe2.dmotinsertionsf_num	ponAbe2.dmotdeletions_num	ponAbe2.dmotdeletionsf_num	ponAbe2.dnoninsertions_num	ponAbe2.dnondeletions_num	ponAbe2.ninsertions_num	ponAbe2.ndeletions_num	ponAbe2.nmotinsertions_num	ponAbe2.nmotinsertionsf_num	ponAbe2.nmotdeletions_num	ponAbe2.nmotdeletionsf_num	ponAbe2.nnoninsertions_num	ponAbe2.nnondeletions_num	ponAbe2.substitutions_num	ponAbe2.bsubstitutions_num	ponAbe2.dsubstitutions_num	ponAbe2.nsubstitutions_num	ponAbe2.indels_num	ponAbe2.subs_num	hg18.insertions_num	hg18.deletions_num	hg18.motinsertions_num	hg18.motinsertionsf_num	hg18.motdeletions_num	hg18.motdeletionsf_num	hg18.noninsertions_num	hg18.nondeletions_num	hg18.binsertions_num	hg18.bdeletions_num	hg18.bmotinsertions_num	hg18.bmotinsertionsf_num	hg18.bmotdeletions_num	hg18.bmotdeletionsf_num	hg18.bnoninsertions_num	hg18.bnondeletions_num	hg18.dinsertions_num	hg18.ddeletions_num	hg18.dmotinsertions_num	hg18.dmotinsertionsf_num	hg18.dmotdeletions_num	hg18.dmotdeletionsf_num	hg18.dnoninsertions_num	hg18.dnondeletions_num	hg18.ninsertions_num	hg18.ndeletions_num	hg18.nmotinsertions_num	hg18.nmotinsertionsf_num	hg18.nmotdeletions_num	hg18.nmotdeletionsf_num	hg18.nnoninsertions_num	hg18.nnondeletions_num	hg18.substitutions_num	hg18.bsubstitutions_num	hg18.dsubstitutions_num	hg18.nsubstitutions_num	hg18.indels_num	hg18.subs_num	panTro2.insertions_num	panTro2.deletions_num	panTro2.motinsertions_num	panTro2.motinsertionsf_num	panTro2.motdeletions_num	panTro2.motdeletionsf_num	panTro2.noninsertions_num	panTro2.nondeletions_num	panTro2.binsertions_num	panTro2.bdeletions_num	panTro2.bmotinsertions_num	panTro2.bmotinsertionsf_num	panTro2.bmotdeletions_num	panTro2.bmotdeletionsf_num	panTro2.bnoninsertions_num	panTro2.bnondeletions_num	panTro2.dinsertions_num	panTro2.ddeletions_num	panTro2.dmotinsertions_num	panTro2.dmotinsertionsf_num	panTro2.dmotdeletions_num	panTro2.dmotdeletionsf_num	panTro2.dnoninsertions_num	panTro2.dnondeletions_num	panTro2.ninsertions_num	panTro2.ndeletions_num	panTro2.nmotinsertions_num	panTro2.nmotinsertionsf_num	panTro2.nmotdeletions_num	panTro2.nmotdeletionsf_num	panTro2.nnoninsertions_num	panTro2.nnondeletions_num	panTro2.substitutions_num	panTro2.bsubstitutions_num	panTro2.dsubstitutions_num	panTro2.nsubstitutions_num	panTro2.indels_num	panTro2.subs_num	insertions_num	deletions_num	motinsertions_num	motinsertionsf_num	motdeletions_num	motdeletionsf_num	noninsertions_num	nondeletions_num	binsertions_num	bdeletions_num	bmotinsertions_num	bmotinsertionsf_num	bmotdeletions_num	bmotdeletionsf_num	bnoninsertions_num	bnondeletions_num	dinsertions_num	ddeletions_num	dmotinsertions_num	dmotinsertionsf_num	dmotdeletions_num	dmotdeletionsf_num	dnoninsertions_num	dnondeletions_num	ninsertions_num	ndeletions_num	nmotinsertions_num	nmotinsertionsf_num	nmotdeletions_num	nmotdeletionsf_num	nnoninsertions_num	nnondeletions_num	substitutions_num	bsubstitutions_num	dsubstitutions_num	nsubstitutions_num	indels_num	subs_num	tree	cleancase' <> '#chrom	start	end	motif	motifsize	threshold	[]	[].chr	[].start	[].end	[].motif	[].motifsize	[].microsat	[].mutation	[].mutation.position	[].mutation.from	[].mutation.to	[].insertion.details	[].deletion.details	somechr	somestart	somened	event	source		insertions_num	deletions_num	motinsertions_num	motinsertionsf_num	motdeletions_num	motdeletionsf_num	noninsertions_num	nondeletions_num	binsertions_num	bdeletions_num	bmotinsertions_num	bmotinsertionsf_num	bmotdeletions_num	bmotdeletionsf_num	bnoninsertions_num	bnondeletions_num	dinsertions_num	ddeletions_num	dmotinsertions_num	dmotinsertionsf_num	dmotdeletions_num	dmotdeletionsf_num	dnoninsertions_num	dnondeletions_num	ninsertions_num	ndeletions_num	nmotinsertions_num	nmotinsertionsf_num	nmotdeletions_num	nmotdeletionsf_num	nnoninsertions_num	nnondeletions_num	substitutions_num	bsubstitutions_num	dsubstitutions_num	nsubstitutions_num	indels_num	subs_num	tree	cleancase'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        221
      </td>
      <td>
        regVar
      </td>
      <td>
        microsats_align1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '5	hg18.chr1	6483	6496	trinucleotide	13	4	GCT	panTro2.chr15	100042575	100042588	trinucleotide	13	4	GCT' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        222
      </td>
      <td>
        regVar
      </td>
      <td>
        microsats_mutability1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        223
      </td>
      <td>
        regVar
      </td>
      <td>
        multispecies_orthologous_microsats
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '>ponAbe2	106	hg18	chr22	16169890	16170045	panTro2	chr22	16191272	16191427	ponAbe2	chr22	13898766	13898941	rheMac2	chr10	61979989	61980162	calJac1	Contig12391	20528	20667	tetranucleotide	ATGT	0	164	+	175	atgtatgtatgt' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        224
      </td>
      <td>
        regVar
      </td>
      <td>
        qualityFilter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/6_quality_filter.maf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        225
      </td>
      <td>
        regVar
      </td>
      <td>
        subRate1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

  	substitution_rates.py 
  	$input 
  	$out_file1
  	#if $region.type == "win":
      ${region.input2} ${region.input2.dbkey} ${region.input2.metadata.chromCol},$region.input2.metadata.startCol,$region.input2.metadata.endCol,$region.input2.metadata.strandCol
    #else:
      "None"
    #end if 
  
Traceback (most recent call last):
  File "substitution_rates.py", line 25, in <module>
    dbkey_i = sys.argv[4]
IndexError: list index out of range

Different lines: '#Seq1	Start1	End1	Seq2	Start2	End2	L	N	p' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        226
      </td>
      <td>
        regVar
      </td>
      <td>
        substitutions1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        227
      </td>
      <td>
        regVar
      </td>
      <td>
        winSplitter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        228
      </td>
      <td>
        rgdat
      </td>
      <td>
        fakePhe1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        229
      </td>
      <td>
        rgdat
      </td>
      <td>
        rgClean1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(25)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        230
      </td>
      <td>
        rgdat
      </td>
      <td>
        rgLDIndep1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(7)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        231
      </td>
      <td>
        rgdat
      </td>
      <td>
        rgPedSub1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        232
      </td>
      <td>
        rgdat
      </td>
      <td>
        rgfakePed1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(5)
      </td>
      <td>
        <pre>Command: 
rgfakePed.py --title '$title'
  -o '$out_file1' -p '$out_file1.files_path' -c '$ncases' -n '$ntotal'
  -s '$nsnp'  -w '$lowmaf' -v '$missingValue' -l '$outFormat'
  -d '$mafdist' -m '$missingRate' -M '$mendelRate' 
Usage: 
Generate either trios or independent subjects with a prespecified
number of random alleles and a uniform or triangular MAF distribution for
stress testing. No LD is simulated - alleles are random. Offspring for
trios are generated by random mating the random parental alleles so there are
no Mendelian errors unless the -M option is used. Mendelian errors are generated
randomly according to the possible errors given the parental mating type although
this is fresh code and not guaranteed to work quite right yet - comments welcomed

Enquiries to ross.lazarus@gmail.com

eg to generate 700 trios with 500k snps, use:
fakebigped.py -n 2100 -s 500000
or to generate 500 independent cases and 500 controls with 100k snps and 0.02 missingness (MCAR), use:
fakebigped.py -c 500 -n 1000 -s 100000 -m 0.02

fakebigped.py -o myfake -m 0.05 -s 100000 -n 2000
will make fbat compatible myfake.ped with 100k markers in
666 trios (total close to 2000 subjects), a uniform MAF distribution and about 5% MCAR missing

fakebigped.py -o myfake -m 0.05 -s 100000 -n 2000 -M 0.05
will make fbat compatible myfake.ped with 100k markers in
666 trios (total close to 2000 subjects), a uniform MAF distribution,
about 5% Mendelian errors and about 5% MCAR missing


fakebigped.py  -o myfakecc -m 0.05 -s 100000 -n 2000 -c 1000 -l
will make plink compatible myfakecc.ped and myfakecc.map (that's what the -l option does),
with 100k markers in 1000 cases and 1000 controls (affection status 2 and 1 respectively),
a triangular MAF distribution (more rare alleles) and about 5% MCAR missing

You should see about 1/4 million genotypes/second so about an hour for a
500k snps in 2k subjects and about a 4GB ped file - these are BIG!!



rgfakePed.py: error: option -c: invalid integer value: "'20'"

File not exists: /tmp/rgtestouts/rgfakePed/rgfakePedtest1.lped
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        233
      </td>
      <td>
        rgdat
      </td>
      <td>
        upload1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        234
      </td>
      <td>
        rgmodel
      </td>
      <td>
        rgCaCo1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        235
      </td>
      <td>
        rgmodel
      </td>
      <td>
        rgGLM1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        236
      </td>
      <td>
        rgmodel
      </td>
      <td>
        rgManQQ1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(60)
      </td>
      <td>
        <pre>Command: 

        rgManQQ.py '$i' "$name" '$out_html' '$out_html.files_path' '$chrom_col' '$offset_col' '$pval_col' '$grey'
    
## Cannot run rgManQQ - missing pval column

File not exists: /tmp/rgtestouts/rgManQQ/rgManQQtest1.html
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        237
      </td>
      <td>
        rgmodel
      </td>
      <td>
        rgTDT1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(79)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        238
      </td>
      <td>
        rgqcplot
      </td>
      <td>
        rgEigPCA1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        239
      </td>
      <td>
        rgqcplot
      </td>
      <td>
        rgGRR1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(350)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        240
      </td>
      <td>
        rgqcplot
      </td>
      <td>
        rgHaploView1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(60)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        241
      </td>
      <td>
        rgqcplot
      </td>
      <td>
        rgManQQ1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(60)
      </td>
      <td>
        <pre>Command: 

        rgManQQ.py '$i' "$name" '$out_html' '$out_html.files_path' '$chrom_col' '$offset_col' '$pval_col' '$grey'
    
## Cannot run rgManQQ - missing pval column

File not exists: /tmp/rgtestouts/rgManQQ/rgManQQtest1.html
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        242
      </td>
      <td>
        rgqcplot
      </td>
      <td>
        rgQC1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(300)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Metadata: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 19, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/Metadata.py", line 7, in main
    ext = sys.argv[1]
IndexError: list index out of range

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        243
      </td>
      <td>
        samtools
      </td>
      <td>
        bam_to_sam
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        3
      </td>
      <td>
        diff(6)
      </td>
      <td>
        <pre>File not exists: /tmp/bam_to_sam_out1.sam
File not exists: /tmp/bam_to_sam_out2.sam
File not exists: /tmp/bam_to_sam_out3.sam
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        244
      </td>
      <td>
        samtools
      </td>
      <td>
        pileup_interval
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        245
      </td>
      <td>
        samtools
      </td>
      <td>
        pileup_parser
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        5001
      </td>
      <td>
        7
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    #if   $pileup_type.type_select == "six"    #pileup_parser.pl $input "3" "5" "6" "4" $qv_cutoff $cvrg_cutoff $snps_only $interval "2" $out_file1 $diff $qc_base
    #elif $pileup_type.type_select == "ten"    #pileup_parser.pl $input "3" "9" "10" "8" $qv_cutoff $cvrg_cutoff $snps_only $interval "2" $out_file1 $diff $qc_base
    #elif $pileup_type.type_select == "manual" #pileup_parser.pl $input $pileup_type.ref_base_column $pileup_type.read_bases_column $pileup_type.read_qv_column $pileup_type.cvrg_column $qv_cutoff $cvrg_cutoff $snps_only $interval $pileup_type.coord_column $out_file1 $diff $qc_base
    #end if#
  
Argument ""3"" isn't numeric in subtraction (-) at pileup_parser.pl line 10.
Argument ""5"" isn't numeric in subtraction (-) at pileup_parser.pl line 11.
Argument ""6"" isn't numeric in subtraction (-) at pileup_parser.pl line 12.
Argument ""4"" isn't numeric in subtraction (-) at pileup_parser.pl line 13.
Argument ""2"" isn't numeric in subtraction (-) at pileup_parser.pl line 18.
Coverage column0 contains non-numeric values. Check your input parameters as well as format of input dataset. at pileup_parser.pl line 40, <IN> line 1.

Different lines: 'chrM	269	270	C	4	a...	;III	1	3	0	0	4' <> 'null'
Command: 

    #if   $pileup_type.type_select == "six"    #pileup_parser.pl $input "3" "5" "6" "4" $qv_cutoff $cvrg_cutoff $snps_only $interval "2" $out_file1 $diff $qc_base
    #elif $pileup_type.type_select == "ten"    #pileup_parser.pl $input "3" "9" "10" "8" $qv_cutoff $cvrg_cutoff $snps_only $interval "2" $out_file1 $diff $qc_base
    #elif $pileup_type.type_select == "manual" #pileup_parser.pl $input $pileup_type.ref_base_column $pileup_type.read_bases_column $pileup_type.read_qv_column $pileup_type.cvrg_column $qv_cutoff $cvrg_cutoff $snps_only $interval $pileup_type.coord_column $out_file1 $diff $qc_base
    #end if#
  
Argument ""3"" isn't numeric in subtraction (-) at pileup_parser.pl line 10.
Argument ""5"" isn't numeric in subtraction (-) at pileup_parser.pl line 11.
Argument ""6"" isn't numeric in subtraction (-) at pileup_parser.pl line 12.
Argument ""4"" isn't numeric in subtraction (-) at pileup_parser.pl line 13.
Argument ""2"" isn't numeric in subtraction (-) at pileup_parser.pl line 18.
Coverage column0 contains non-numeric values. Check your input parameters as well as format of input dataset. at pileup_parser.pl line 40, <IN> line 1.

Different lines: 'chrM	270	C	4	a...	;III	1	3	0	0	4' <> 'null'
Command: 

    #if   $pileup_type.type_select == "six"    #pileup_parser.pl $input "3" "5" "6" "4" $qv_cutoff $cvrg_cutoff $snps_only $interval "2" $out_file1 $diff $qc_base
    #elif $pileup_type.type_select == "ten"    #pileup_parser.pl $input "3" "9" "10" "8" $qv_cutoff $cvrg_cutoff $snps_only $interval "2" $out_file1 $diff $qc_base
    #elif $pileup_type.type_select == "manual" #pileup_parser.pl $input $pileup_type.ref_base_column $pileup_type.read_bases_column $pileup_type.read_qv_column $pileup_type.cvrg_column $qv_cutoff $cvrg_cutoff $snps_only $interval $pileup_type.coord_column $out_file1 $diff $qc_base
    #end if#
  
Argument ""3"" isn't numeric in subtraction (-) at pileup_parser.pl line 10.
Argument ""5"" isn't numeric in subtraction (-) at pileup_parser.pl line 11.
Argument ""6"" isn't numeric in subtraction (-) at pileup_parser.pl line 12.
Argument ""4"" isn't numeric in subtraction (-) at pileup_parser.pl line 13.
Argument ""2"" isn't numeric in subtraction (-) at pileup_parser.pl line 18.
Coverage column0 contains non-numeric values. Check your input parameters as well as format of input dataset. at pileup_parser.pl line 40, <IN> line 1.

Different lines: 'chrM	45	A	3	..^:.	III	3	0	0	0	3' <> 'null'
Command: 

    #if   $pileup_type.type_select == "six"    #pileup_parser.pl $input "3" "5" "6" "4" $qv_cutoff $cvrg_cutoff $snps_only $interval "2" $out_file1 $diff $qc_base
    #elif $pileup_type.type_select == "ten"    #pileup_parser.pl $input "3" "9" "10" "8" $qv_cutoff $cvrg_cutoff $snps_only $interval "2" $out_file1 $diff $qc_base
    #elif $pileup_type.type_select == "manual" #pileup_parser.pl $input $pileup_type.ref_base_column $pileup_type.read_bases_column $pileup_type.read_qv_column $pileup_type.cvrg_column $qv_cutoff $cvrg_cutoff $snps_only $interval $pileup_type.coord_column $out_file1 $diff $qc_base
    #end if#
  
Argument ""3"" isn't numeric in subtraction (-) at pileup_parser.pl line 10.
Argument ""9"" isn't numeric in subtraction (-) at pileup_parser.pl line 11.
Argument ""10"" isn't numeric in subtraction (-) at pileup_parser.pl line 12.
Argument ""8"" isn't numeric in subtraction (-) at pileup_parser.pl line 13.
Argument ""2"" isn't numeric in subtraction (-) at pileup_parser.pl line 18.
Coverage column0 contains non-numeric values. Check your input parameters as well as format of input dataset. at pileup_parser.pl line 40, <IN> line 1.

Different lines: 'chrM	13	14	A	A	56	0	25	18	.......G.........^:.	BIIIIIII+IIIIIIIII	16	0	1	0	17' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        246
      </td>
      <td>
        samtools
      </td>
      <td>
        sam2interval
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'ref	6	22	+	r001	163	ref	7	30	8M2I4M1D3M	=	37	39	TTAGATAAAGGATACTA	*' <> 'ref	6	22	+'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        247
      </td>
      <td>
        samtools
      </td>
      <td>
        sam_bw_filter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 98, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'states'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        248
      </td>
      <td>
        samtools
      </td>
      <td>
        sam_merge
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/sam_merge_out2.bam
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        249
      </td>
      <td>
        samtools
      </td>
      <td>
        sam_pileup
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 110, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'metadata'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 110, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'metadata'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        250
      </td>
      <td>
        samtools
      </td>
      <td>
        sam_to_bam
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 94, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'metadata'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 104, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'metadata'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        251
      </td>
      <td>
        samtools
      </td>
      <td>
        samtools_flagstat
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/samtools_flagstat_out1.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        252
      </td>
      <td>
        scores
      </td>
      <td>
        aggregate_scores_in_intervals2
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    #if $score_source_type.score_source == "user" #aggregate_scores_in_intervals.py $score_source_type.input2 $input1 ${input1.metadata.chromCol} ${input1.metadata.startCol} ${input1.metadata.endCol} $out_file1 --chrom_buffer=3
    #else                                         #aggregate_scores_in_intervals.py $score_source_type.datasets $input1 ${input1.metadata.chromCol} ${input1.metadata.startCol} ${input1.metadata.endCol} $out_file1 -b
    #end if#
  
Traceback (most recent call last):
  File "aggregate_scores_in_intervals.py", line 25, in <module>
    from bx.binned_array import BinnedArray, FileBinnedArray
  File "/usr/local/lib/python2.7/dist-packages/bx_python-0.7.0_494c2d1d68b3_rebuild1-py2.7-linux-x86_64-ucs4.egg/bx/binned_array.py", line 15, in <module>
    from numpy import *
ImportError: No module named numpy

File not exists: /tmp/aggregate_binned_scores_in_intervals.out
Command: 

    #if $score_source_type.score_source == "user" #aggregate_scores_in_intervals.py $score_source_type.input2 $input1 ${input1.metadata.chromCol} ${input1.metadata.startCol} ${input1.metadata.endCol} $out_file1 --chrom_buffer=3
    #else                                         #aggregate_scores_in_intervals.py $score_source_type.datasets $input1 ${input1.metadata.chromCol} ${input1.metadata.startCol} ${input1.metadata.endCol} $out_file1 -b
    #end if#
  
Traceback (most recent call last):
  File "aggregate_scores_in_intervals.py", line 25, in <module>
    from bx.binned_array import BinnedArray, FileBinnedArray
  File "/usr/local/lib/python2.7/dist-packages/bx_python-0.7.0_494c2d1d68b3_rebuild1-py2.7-linux-x86_64-ucs4.egg/bx/binned_array.py", line 15, in <module>
    from numpy import *
ImportError: No module named numpy

File not exists: /tmp/aggregate_binned_scores_in_intervals2.interval
Command: 

    #if $score_source_type.score_source == "user" #aggregate_scores_in_intervals.py $score_source_type.input2 $input1 ${input1.metadata.chromCol} ${input1.metadata.startCol} ${input1.metadata.endCol} $out_file1 --chrom_buffer=3
    #else                                         #aggregate_scores_in_intervals.py $score_source_type.datasets $input1 ${input1.metadata.chromCol} ${input1.metadata.startCol} ${input1.metadata.endCol} $out_file1 -b
    #end if#
  
Traceback (most recent call last):
  File "aggregate_scores_in_intervals.py", line 25, in <module>
    from bx.binned_array import BinnedArray, FileBinnedArray
  File "/usr/local/lib/python2.7/dist-packages/bx_python-0.7.0_494c2d1d68b3_rebuild1-py2.7-linux-x86_64-ucs4.egg/bx/binned_array.py", line 15, in <module>
    from numpy import *
ImportError: No module named numpy

File not exists: /tmp/aggregate_binned_scores_in_intervals3.out
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        253
      </td>
      <td>
        scores
      </td>
      <td>
        phastOdds_for_intervals
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
get_scores_galaxy.py $per_col ${score_file}.h5 ${score_file}.mapping.bed $input $output ${input.metadata.chromCol} ${input.metadata.startCol} ${input.metadata.endCol}
Traceback (most recent call last):
  File "get_scores_galaxy.py", line 12, in <module>
    from numpy import *
ImportError: No module named numpy

File not exists: /tmp/phastOdds_tool_out.interval
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        254
      </td>
      <td>
        scores
      </td>
      <td>
        wiggle2simple1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        255
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        PerM
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/perm_out1.sam
File not exists: /tmp/perm_out3.fastqsanger
File not exists: /tmp/perm_out2.sam
File not exists: /tmp/perm_out4.sam
File not exists: /tmp/perm_out5.sam
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        256
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        bfast_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        4
      </td>
      <td>
        diff(2)
      </td>
      <td>
        <pre>Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 153, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'fields'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        257
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        bowtie_color_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        258
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        bowtie_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        6
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        259
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        bwa_color_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        4
      </td>
      <td>
        diff(2)
      </td>
      <td>
        <pre>File not exists: /tmp/bwa_wrapper_out4.sam
Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 111, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 238, in valueForName
    _wrapNotFoundException(e, fullName=name, namespace=obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 236, in valueForName
    return _valueForName(obj, name, executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'tool_data_tables' while searching for 'tool_data_tables'

'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        260
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        bwa_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 110, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 238, in valueForName
    _wrapNotFoundException(e, fullName=name, namespace=obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 236, in valueForName
    return _valueForName(obj, name, executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'tool_data_tables' while searching for 'tool_data_tables'

'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'Error: java.lang.NullPointerException
'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        261
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        lastz_paired_reads_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 122, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 238, in valueForName
    _wrapNotFoundException(e, fullName=name, namespace=obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 236, in valueForName
    return _valueForName(obj, name, executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'tool_data_tables' while searching for 'tool_data_tables'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        262
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        lastz_wrapper_2
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 122, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 238, in valueForName
    _wrapNotFoundException(e, fullName=name, namespace=obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 236, in valueForName
    return _valueForName(obj, name, executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'tool_data_tables' while searching for 'tool_data_tables'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 116, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'metadata'

'Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 116, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 245, in valueFromSearchList
    executeCallables=executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'metadata'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        263
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        megablast_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 88, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 238, in valueForName
    _wrapNotFoundException(e, fullName=name, namespace=obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 236, in valueForName
    return _valueForName(obj, name, executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'tool_data_tables' while searching for 'tool_data_tables'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        264
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        megablast_xml_parser
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/megablast_xml_parser_test1_out.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        265
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        mosaik_wrapper
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        sim_size(0)
      </td>
      <td>
        <pre>File not exists: /tmp/mosaik_test_out.sam
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        266
      </td>
      <td>
        solexa_tools
      </td>
      <td>
        srma_wrapper
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(2)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 95, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 238, in valueForName
    _wrapNotFoundException(e, fullName=name, namespace=obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 236, in valueForName
    return _valueForName(obj, name, executeCallables)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 226, in _valueForName
    _raiseNotFoundException(key, obj)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'tool_data_tables' while searching for 'tool_data_tables'

'File not exists: /tmp/srma_out2.bam
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        267
      </td>
      <td>
        stats
      </td>
      <td>
        Count1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/uniq_out.dat
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        268
      </td>
      <td>
        stats
      </td>
      <td>
        Summary_Statistics1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
gsummary.py $input $out_file1 "$cond"
Traceback (most recent call last):
  File "gsummary.py", line 4, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/gsummary_out1.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        269
      </td>
      <td>
        stats
      </td>
      <td>
        compute_q_values
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Error: java.lang.Exception: Fill command template: /home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module Cheetah was already imported from /usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/Cheetah-2.2.2-py2.7-linux-x86_64-ucs4.egg is being added to sys.path
  self.check_version_conflict()
/home/tolstyh/projects/other/galaxy/galaxy-dist/lib/galaxy/__init__.py:79: UserWarning: Module sqlalchemy was already imported from /usr/local/lib/python2.7/dist-packages/SQLAlchemy-0.7.2-py2.7.egg/sqlalchemy/__init__.pyc, but /home/tolstyh/projects/other/galaxy/galaxy-dist/eggs/SQLAlchemy-0.5.6_dev_r6498-py2.7.egg is being added to sys.path
  self.check_version_conflict()
/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Compiler.py:1509: UserWarning: 
You don't have the C version of NameMapper installed! I'm disabling Cheetah's useStackFrames option as it is painfully slow with the Python version of NameMapper. You should get a copy of Cheetah with the compiled C version of NameMapper.
  "\nYou don't have the C version of NameMapper installed! "
Traceback (most recent call last):
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 71, in <module>
    if __name__ == "__main__":main()
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 43, in main
    print fillTemplate(template, params)
  File "/home/tolstyh/projects/java/BioUML/src/../plugins/ru.biosoft.galaxy_0.9.2/FillTemplate.py", line 69, in fillTemplate
    return str( Template( source=template, searchList=[params] ))
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/Template.py", line 1005, in __str__
    rc = getattr(self, mainMethName)()
  File "DynamicallyCompiledCheetahTemplate.py", line 105, in respond
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 246, in valueFromSearchList
    _raiseNotFoundException(key, searchList)
  File "/usr/local/lib/python2.7/dist-packages/Cheetah-2.4.4-py2.7.egg/Cheetah/NameMapper.py", line 167, in _raiseNotFoundException
    raise NotFound(excString)
Cheetah.NameMapper.NotFound: cannot find 'outputFile2'

'</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        270
      </td>
      <td>
        stats
      </td>
      <td>
        cor2
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
cor.py $input1 $out_file1 $numeric_columns $method
Traceback (most recent call last):
  File "cor.py", line 9, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/cor_out.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        271
      </td>
      <td>
        stats
      </td>
      <td>
        count_gff_features
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        272
      </td>
      <td>
        stats
      </td>
      <td>
        generate_matrix_for_pca_and_lda1
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        273
      </td>
      <td>
        stats
      </td>
      <td>
        lda_analy1
      </td>
      <td>
        sh&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/lda_analy_output.txt
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        274
      </td>
      <td>
        stats
      </td>
      <td>
        plot_for_lda_output1
      </td>
      <td>
        sh&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/plot_for_lda_output.pdf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        275
      </td>
      <td>
        stats
      </td>
      <td>
        t_test_two_samples
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        4001
      </td>
      <td>
        6
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: '' <> 'null'
Different lines: '' <> 'null'
Command: 

  	t_test_two_samples.pl $inputFile1 $inputFile2 $inputTestSidedness3 $inputStandardDeviationEquality4 $outputFile1
  
usage: non_pooled_t_test_two_samples_galaxy.pl [TABULAR.in] [TABULAR.in] [testSidedness] [standardDeviationEquality] [TXT.out] 

File not exists: /tmp/t_test_result3.text
Command: 

  	t_test_two_samples.pl $inputFile1 $inputFile2 $inputTestSidedness3 $inputStandardDeviationEquality4 $outputFile1
  
usage: non_pooled_t_test_two_samples_galaxy.pl [TABULAR.in] [TABULAR.in] [testSidedness] [standardDeviationEquality] [TXT.out] 

File not exists: /tmp/t_test_result4.text
Command: 

  	t_test_two_samples.pl $inputFile1 $inputFile2 $inputTestSidedness3 $inputStandardDeviationEquality4 $outputFile1
  
usage: non_pooled_t_test_two_samples_galaxy.pl [TABULAR.in] [TABULAR.in] [testSidedness] [standardDeviationEquality] [TXT.out] 

File not exists: /tmp/t_test_result5.text
Command: 

  	t_test_two_samples.pl $inputFile1 $inputFile2 $inputTestSidedness3 $inputStandardDeviationEquality4 $outputFile1
  
usage: non_pooled_t_test_two_samples_galaxy.pl [TABULAR.in] [TABULAR.in] [testSidedness] [standardDeviationEquality] [TXT.out] 

File not exists: /tmp/t_test_result6.text
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        276
      </td>
      <td>
        tax_manipulation
      </td>
      <td>
        Draw_phylogram
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        0
      </td>
      <td>
        null
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        277
      </td>
      <td>
        tax_manipulation
      </td>
      <td>
        Fetch Taxonomic Ranks
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
gi2taxonomy.py $input $giField $idField $out_file1 ${GALAXY_DATA_INDEX_DIR}
Traceback (most recent call last):
  File "gi2taxonomy.py", line 166, in <module>
    get_taxId( GI2TAX, collapse_repeating_gis( g2n ), tb_F.name )
  File "gi2taxonomy.py", line 109, in get_taxId
    for line in file( gi2tax_file ):
IOError: [Errno 2] No such file or directory: '/home/tolstyh/projects/other/galaxy/galaxy-dist/tool-data/taxonomy/gi_taxid_sorted.txt'

File not exists: /tmp/taxonomy2gi-output.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        278
      </td>
      <td>
        tax_manipulation
      </td>
      <td>
        find_diag_hits
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        279
      </td>
      <td>
        tax_manipulation
      </td>
      <td>
        lca1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        280
      </td>
      <td>
        tax_manipulation
      </td>
      <td>
        poisson2test
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
poisson2test.py $input1 $input2 $input3 $input4 $input5 $output1 2>/dev/null 
Traceback (most recent call last):
  File "poisson2test.py", line 5, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/poisson2test1_out.tabular
Command: 
poisson2test.py $input1 $input2 $input3 $input4 $input5 $output1 2>/dev/null 
Traceback (most recent call last):
  File "poisson2test.py", line 5, in <module>
    from rpy import *
ImportError: No module named rpy

File not exists: /tmp/poisson2test2_out.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        281
      </td>
      <td>
        tax_manipulation
      </td>
      <td>
        t2t_report
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/t2t_report.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        282
      </td>
      <td>
        textutil
      </td>
      <td>
        Add_a_column1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        5001
      </td>
      <td>
        3
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/column_maker_out1.interval
File not exists: /tmp/column_maker_out2.interval
File not exists: /tmp/column_maker_out3.interval
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        283
      </td>
      <td>
        textutil
      </td>
      <td>
        ChangeCase
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 
changeCase.pl $input "$cols" $delimiter $casing $out_file1
No columns specified, columns are not preceeded with 'c', or commas are not used to separate column numbers: "c1"

File not exists: /tmp/changeCase_out1.tabular
Command: 
changeCase.pl $input "$cols" $delimiter $casing $out_file1
No columns specified, columns are not preceeded with 'c', or commas are not used to separate column numbers: "c1"

File not exists: /tmp/changeCase_out2.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        284
      </td>
      <td>
        textutil
      </td>
      <td>
        Convert characters1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        285
      </td>
      <td>
        textutil
      </td>
      <td>
        Cut1
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr1	CCDS989.1_cds_0_0_chr1_147962193_r	147962192	147962580' <> 'CCDS989.1_cds_0_0_chr1_147962193_r	147962192'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        286
      </td>
      <td>
        textutil
      </td>
      <td>
        Paste1
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        287
      </td>
      <td>
        textutil
      </td>
      <td>
        Remove beginning1
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        288
      </td>
      <td>
        textutil
      </td>
      <td>
        Show beginning1
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        289
      </td>
      <td>
        textutil
      </td>
      <td>
        Show tail1
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        290
      </td>
      <td>
        textutil
      </td>
      <td>
        addValue
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1001
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        291
      </td>
      <td>
        textutil
      </td>
      <td>
        cat1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        292
      </td>
      <td>
        textutil
      </td>
      <td>
        createInterval
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        293
      </td>
      <td>
        textutil
      </td>
      <td>
        dna_filter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        3001
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Different lines: 'chr1	582	583	G	G	G	N	G	R	N	R	N	-	N	M	N	V	K	N	N	N	G	C	R	+' <> 'null'
File not exists: /tmp/dna_filter_out2.tabular
Different lines: 'chr1	468	469	C	C	C	N	M	N	N	K	.	N	C	U	N	H	N	G	N	N	M	N	S	-' <> 'null'
Different lines: 'chr2	45788	45789	T	T	T	N	W	S	N	Y	N	R	Y	N	S	N	W	M	N	C	T	N	C	-' <> 'null'
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        294
      </td>
      <td>
        textutil
      </td>
      <td>
        mergeCols1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        295
      </td>
      <td>
        textutil
      </td>
      <td>
        random_lines1
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        1000
      </td>
      <td>
        1
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        296
      </td>
      <td>
        textutil
      </td>
      <td>
        tables_arithmetic_operations
      </td>
      <td>
        perl&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        3001
      </td>
      <td>
        4
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        297
      </td>
      <td>
        textutil
      </td>
      <td>
        trimmer
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        298
      </td>
      <td>
        textutil
      </td>
      <td>
        wc_gnu
      </td>
      <td>
        &nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        1001
      </td>
      <td>
        2
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>File not exists: /tmp/wc_gnu_out_1.tabular
File not exists: /tmp/wc_gnu_out_2.tabular
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        299
      </td>
      <td>
        vcf_tools
      </td>
      <td>
        vcf_annotate
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2001
      </td>
      <td>
        2
      </td>
      <td>
        diff(6)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        300
      </td>
      <td>
        vcf_tools
      </td>
      <td>
        vcf_extract
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='red'>error</font>
      </td>
      <td>
        4001
      </td>
      <td>
        5
      </td>
      <td>
        diff(0)
      </td>
      <td>
        <pre>Command: 

    vcfPytools.py
      extract 
      --in=$input1
      --out=$output1
      #if $reference_sequence.value.strip()
        --reference-sequence=$reference_sequence
      #end if
      #if $region.value.strip()
        --region=$region
      #end if
      #if $keep_quality.value.strip()
        --keep-quality=$keep_quality
      #end if
      #if $keep_info.value.strip()
      --keep-info=$keep_info
      #end if
      #if $discard_info.value.strip()
        --discard-info=$discard_info
      #end if
      $pass_filter
  

A region (--region, -r) or reference sequence (--reference-sequence, -s) must be supplied
if not extracting records based on info strings.

File not exists: /tmp/test_extract_pass_filter_quality_9_DP_2000_lt.vcf
</pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        301
      </td>
      <td>
        vcf_tools
      </td>
      <td>
        vcf_filter
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        2000
      </td>
      <td>
        2
      </td>
      <td>
        diff(6)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
    <tr>
      <td>
        302
      </td>
      <td>
        vcf_tools
      </td>
      <td>
        vcf_intersect
      </td>
      <td>
        python&nbsp;
      </td>
      <td>
        <font color='green'>OK</font>
      </td>
      <td>
        4001
      </td>
      <td>
        3
      </td>
      <td>
        diff(2)
      </td>
      <td>
        <pre></pre>&nbsp;
      </td>
    </tr>
  </table>
</body>
</html>
